{{YmlParametersStart}}

{{YmlParametersEnd}}

---
date: "`r paste0(Sys.time(), ' (einprot v', packageVersion('einprot'), ')')`"
output: 
    html_document:
        theme: united
        toc: true
        toc_float: true
        code_folding: hide
editor_options: 
  chunk_output_type: console
bibliography: "`r system.file('extdata', 'einprot_bibliography.bib', package = 'einprot')`"
---

{{ConfigParametersStart}}

{{ConfigParametersEnd}}

---
title: "`r reportTitle`"
author: "`r reportAuthor`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, dpi = 200)
```

```{r load-pkg, echo=FALSE}
## Load packages
suppressPackageStartupMessages({
    library(QFeatures)
    library(SummarizedExperiment)
    library(SingleCellExperiment)
    library(imputeLCMD)
    library(rrcovNA)
    library(scater)
    library(dplyr)
    library(tibble)
    library(tidyr)
    library(ggplot2)
    library(ggrepel)
    library(cowplot)
    library(plotly)
    library(UpSetR)
    library(circlize)
    library(ComplexHeatmap)
    library(DT)
    library(STRINGdb)
    library(MsCoreUtils)
    library(ComplexUpset)
    library(ExploreModelMatrix)
    library(einprot)
})
```

This report performs PTM analysis by investigating differences in peptide 
intensities relative to differences in the corrresponding protein. 
Most of the code is hidden by default, 
but can be displayed by clicking on the `Code` buttons (or by selecting 
`Code -> Show All Code` in the top right corner of the report).

```{r get-basic-info}
## Get species info
speciesInfo <- getSpeciesInfo(species)

## Get conversion tables for PomBase and WormBase IDs
pbconv <- readRDS(system.file(EINPROT_POMBASE_CONVTABLE,
                              package = "einprot"))
wbconv <- readRDS(system.file(EINPROT_WORMBASE_CONVTABLE,
                              package = "einprot"))
```

# Experiment details

```{r exp-table}
makeTableFromList(c(experimentInfo, 
                    list(
                        "Species" = speciesInfo$species,
                        "Species (common)" = speciesInfo$speciesCommon,
                        "Taxonomic ID" = speciesInfo$taxId
                    )))
```

# Input files

```{r inp-table}
makeTableFromList(list(
    "SCE with protein data" = sceProteins,
    "SCE with peptide data" = scePeptides
))
```

# Settings {#settings-table}

```{r settings-table}
settingsList <- list(
    "Assay to use for tests" = assayForTests,
    "Assay with imputation information" = assayImputation,
    "Min. nbr valid values" = minNbrValidValues,
    "Model fit" = ifelse(singleFit, "Single (one model fit for all samples)", 
                         "Separate model fit for each comparison"),
    "Comparisons" = paste(unlist(lapply(comparisons, 
                                        function(x) paste(x, collapse = " vs "))),
                          collapse = "; "),
    "Control group" = ctrlGroup,
    "Do all pairwise comparisons" = allPairwiseComparisons,
    "Statistical testing approach" = testType,
    "Minimal fold change (limma/treat)" = minlFC,
    "Adjusted p-value threshold for volcano plots" = volcanoAdjPvalThr,
    "Log2 FC threshold for volcano plots" = volcanoLog2FCThr,
    "Max nbr features to indicate in volcano plots" = volcanoMaxFeatures,
    "Features to always label in volcano plots" = paste(volcanoFeaturesToLabel,
                                                        collapse = ", "),
    "Random seed" = seed)

makeTableFromList(settingsList)
```

# Read processed data

The input to this workflow is two SingleCellExperiment objects, containing 
peptide and protein intensities, respectively. 
First, if requested, we filter out any rows corresponding to unmodified 
peptides (where 'Modifications.in.Master.Proteins' is empty) from the peptide 
data.
For the PTM test, we'll need to find the corresponding protein for each peptide. 
We start by generating a matching column in each of the data sets. 
Peptides without a corresponding protein record will not be used for the 
analyses, but will be included in the final object. 


```{r read-filter-data}
## Read data
sceProteins <- readRDS(sceProteins)
scePeptides <- readRDS(scePeptides)

## Exclude unmodified peptides
if (!modificationsCol %in% colnames(rowData(scePeptides))) {
    stop("The peptides SCE does not have a column named ", modificationsCol)
}
if (excludeUnmodifiedPeptides) {
    scePeptides <- scePeptides[rowData(scePeptides)[[modificationsCol]] != "", ]
}
if (!is.null(keepModifications)) {
    scePeptides <- scePeptides[grep(keepModifications, rowData(scePeptides)[[modificationsCol]]), ]
}
if (nrow(scePeptides) == 0) {
    stop("No peptides left after filtering for modifications")
}

## Generate match column
scePeptides <- fixFeatureIds(scePeptides, 
                             colDefs = list(einprotMatchProt = proteinIdColPeptides))
sceProteins <- fixFeatureIds(sceProteins, 
                             colDefs = list(einprotMatchProt = proteinIdColProteins))

makeTableFromList(list(
    "Number of shared proteins" = length(
        intersect(rowData(scePeptides)[["einprotMatchProt"]], 
                  rowData(sceProteins)[["einprotMatchProt"]])),
    "Number of shared samples" = length(
        intersect(colnames(scePeptides), colnames(sceProteins)))
))
```

# Check sample annotations

Each of the SingleCellExperiment objects contains sample annotations. 

## Proteins

```{r get-sampleinfo-proteins}
DT::datatable(as.data.frame(colData(sceProteins)),
              options = list(scrollX = TRUE, pageLength = 20))
```

## Peptides

```{r get-sampleinfo-peptides}
DT::datatable(as.data.frame(colData(scePeptides)),
              options = list(scrollX = TRUE, pageLength = 20))
```

# Overall distribution of intensities

The box plot below displays the distribution of protein and peptide intensities 
in each sample. These values will be used to perform the PTM test below.

```{r intensity-distribution, warning=FALSE, message=FALSE, fig.width = min(14, max(7, 0.5 * ncol(sceProteins))), fig.height = 5/7 * min(14, max(7, 0.5 * ncol(sceProteins)))}
makeIntensityBoxplots(sce = sceProteins, assayName = assayForTests, doLog = FALSE, 
                      ylab = "Protein intensity")
makeIntensityBoxplots(sce = scePeptides, assayName = assayForTests, doLog = FALSE, 
                      ylab = "Peptide intensity")
```

# Statistical testing

For each feature, we then compare the peptide log2 intensities between groups, 
relative to the change in protein intensities. This is done using `limma`. 

## Comparisons and design

The `r assayForTests` assay will be used for the tests. The 
following pairwise comparisons will be performed (in each case, the 
first listed group will be the 'baseline' group): 

```{r initialize-tests}
(comparisonList <- makeListOfComparisons(
    allGroups = unique(sceProteins$group), comparisons = comparisons, 
    mergeGroups = list(),
    allPairwiseComparisons = allPairwiseComparisons,
    ctrlGroup = ctrlGroup, discardGroup = NULL))
```

```{r run-test}
testres <- runPTMTest(sceProteins = sceProteins, 
                      scePeptides = scePeptides, 
                      matchColProteins = "einprotMatchProt",
                      matchColPeptides = "einprotMatchProt",
                      testType = testType,
                      comparisons = comparisonList$comparisons, 
                      groupComposition = comparisonList$groupComposition, 
                      assayForTests = assayForTests, 
                      assayImputation = assayImputation, 
                      minNbrValidValues = minNbrValidValues, 
                      minlFC = minlFC, 
                      volcanoAdjPvalThr = volcanoAdjPvalThr, 
                      volcanoLog2FCThr = volcanoLog2FCThr, 
                      singleFit = singleFit, baseFileName = NULL, 
                      subtractBaseline = FALSE, 
                      baselineGroup = "")
tests <- testres$tests
```


```{r text-expdesign, results="asis", echo=FALSE}
cat(expDesignText(testType = "limma"))
```

```{r get-fig-height, echo = FALSE}
if ("design" %in% names(testres$design)) {
    ## Overall fit
    ngroups <- length(unique(testres$design$sampleData$group))
    fh <- max(4, 1.5 * ngroups)
} else {
    ## Separate fits - always two groups
    fh <- 4
}
```

```{r expdesign-plot, eval = (testType == "interaction"), echo = (testType == "interaction"), fig.width = 12, fig.height = 2 * fh, results = "asis"}
if ("design" %in% names(testres$design)) {
    cat("\n### Overall design \n")
    vd <- VisualizeDesign(
        testres$design$sampleData[, c("group", "sample", "dataLevel")], designFormula = NULL, 
        designMatrix = testres$design$design, textSizeFitted = 4)
        print(cowplot::plot_grid(
            cowplot::plot_grid(plotlist = vd$plotlist, ncol = 1), 
            cowplot::plot_grid(plotlist = vd$cooccurrenceplots, ncol = 1), 
            ncol = 2
        ))
    cat("\n\n")
    cat("<details>\n<summary><b>\nClick to display design matrix and contrast(s)\n</b></summary>\n")
    cat("\n````\n")
    print(testres$design$design)
    cat("\n````\n")
    cat("\n````\n")
    print(testres$design$contrasts)
    cat("\n````\n")
    cat("\n</details>\n\n")
} else {
    for (nm in names(testres$design)) {
        cat("\n###", nm, "\n")
        vd <- VisualizeDesign(
            testres$design[[nm]]$sampleData[, c("group", "sample", "dataLevel")], 
            designFormula = NULL, 
            designMatrix = testres$design[[nm]]$design, textSizeFitted = 4)
        print(cowplot::plot_grid(
            cowplot::plot_grid(plotlist = vd$plotlist, ncol = 1), 
            cowplot::plot_grid(plotlist = vd$cooccurrenceplots, ncol = 1), 
            ncol = 2
        ))
        cat("\n\n")
        cat("<details>\n<summary><b>\nClick to display design matrix and contrast(s)\n</b></summary>\n")
        cat("\n````\n")
        print(testres$design[[nm]]$design)
        cat("\n````\n")
        cat("\n````\n")
        print(testres$design[[nm]]$contrast)
        cat("\n````\n")
        cat("\n</details>\n\n")
    }
}
```

```{r test-messages, echo = (length(testres$messages) > 0), eval = (length(testres$messages) > 0)}
testres$messages
```


```{r text-sa, results="asis", echo=FALSE, eval = (testType == "interaction")}
cat("\n## SA plots \n")
cat(saText(testType = "limma"))
```

```{r plot-sa, fig.width = 12, fig.height = ceiling(length(testres$tests)/3) * 4, warning = FALSE, eval = (testType == "interaction"), echo = (testType == "interaction")}
makeSAPlot(testList = testres$tests)
```

## Volcano plots

Below we display a volcano plot for each comparison. These plots are also 
saved to pdf files. In each plot, a subset of the significant hits are 
indicated by name (n=`r volcanoMaxFeatures`, selected as the ones with 
the largest Manhattan distance to the origin). Any features 
explicitly requested (see the [table above](#settings-table)) are also labeled
in the volcano plots. 

```{r volcano-plot, warning=FALSE, fig.width=10.5, fig.height=7.5, results="asis"}
interactiveVolcanos <- htmltools::tagList()
for (nm in names(testres$tests)) {
    plots <- plotVolcano(
        sce = scePeptides, res = testres$tests[[nm]], testType = testType, 
        xv = NULL, yv = NULL, xvma = NULL, volcind = NULL, 
        plotnote = testres$plotnotes[[nm]], 
        plottitle = testres$plottitles[[nm]], 
        plotsubtitle = testres$plotsubtitles[[nm]],
        volcanoFeaturesToLabel = volcanoFeaturesToLabel, 
        volcanoMaxFeatures = volcanoMaxFeatures,
        baseFileName = paste0(sub("\\.Rmd$", "", knitr::current_input()),
                              "_volcano_", nm), 
        comparisonString = nm, 
        groupComposition = comparisonList$groupComposition[comparisonList$comparisons[[nm]]],
        stringDb = NULL,
        curveparam = list(),
        abundanceColPat = assayForTests,
        xlab = "log2(fold change)", 
        ylab = "-log10(p-value)",
        xlabma = "Average abundance",
        labelOnlySignificant = TRUE, 
        interactiveDisplayColumns = interactiveDisplayColumns)
    if (!is.null(plots$ggma) && !is.null(plots$ggwf)) {
        cat("\n\n### ", nm, " \n\n\n")
        print(plots$gg)
        print(plots$ggma)
        print(plots$ggwf)
        cat("\n\n")
    } else if (!is.null(plots$ggma)) {
        cat("\n\n### ", nm, " \n\n\n")
        print(plots$gg)
        print(plots$ggma)
        cat("\n\n")
    } else if (!is.null(plots$ggwf)) {
        cat("\n\n### ", nm, " \n\n\n")
        print(plots$gg)
        print(plots$ggwf)
        cat("\n\n")
    } else {
        cat("\n\n### ", nm, " \n\n\n")
        print(plots$gg)
        cat("\n\n")
    }
    interactiveVolcanos[[nm]] <- plots$ggint
}
```

```{r interactive-volcanos-text, echo = FALSE, eval = addInteractiveVolcanos, results = "asis"}
cat("## Interactive volcano plots")
```

```{r interactive-volcanos, echo = FALSE, eval = addInteractiveVolcanos, fig.width = 10.5, fig.height = 8}
interactiveVolcanos
```

## Result export

For each comparison, we also save a text file with the "significant" features 
(defined as those colored in the volcano plots above). The features are 
ordered by the logFC value. 

```{r merge-tests}
## Merge results from all tests and add to rowData(sce)
tests <- testres$tests
for (nm in names(tests)) {
    idx <- which(colnames(tests[[nm]]) != "pid")
    colnames(tests[[nm]])[idx] <- paste0(nm, ".", colnames(tests[[nm]])[idx])
}
all_tests <- as.data.frame(Reduce(function(...) dplyr::full_join(..., by = "pid"),
                                  tests), optional = TRUE)
rownames(all_tests) <- all_tests$pid
all_tests$pid <- NULL
colnames(all_tests) <- paste0("PTMs_", colnames(all_tests))

colnames(rowData(scePeptides)) <- paste0("Peptides_", colnames(rowData(scePeptides)))

stopifnot(rownames(scePeptides) == rownames(all_tests))
rowData(scePeptides) <- cbind(rowData(scePeptides), all_tests)
```

## Overlap among sets of significant features

The UpSet plot below shows the overlap among the "significant" proteins 
(defined as the ones that are colored in the volcano plots above, based 
on the user specifications) from the different comparisons. Note that if 
there are many comparisons, not all combinations may be displayed in 
the plot (only the 50 combinations with the largest number of proteins 
are shown, for interpretability reasons). Moreover, the UpSet plot will 
only be shown if at least two comparisons have been made, and there are 
at least two comparisons where any proteins were deemed significant.

```{r upset-tests, fig.width = 9}
tmpsign <- all_tests %>% dplyr::select(contains("showInVolcano")) + 0
tmpsign[is.na(tmpsign)] <- 0
colnames(tmpsign) <- gsub("\\.showInVolcano$", "", colnames(tmpsign))
if (length(tests) > 1 && sum(colSums(tmpsign) > 0) > 1) {
    UpSetR::upset(tmpsign, nsets = length(tests), nintersects = 50, 
                  order.by = "freq")
}
```

# Table with direct database links to sequences, functional information and predicted structures

The table below provides autogenerated links to the UniProt and 
AlphaFold pages (as well as selected organism-specific databases) for the 
protein IDs corresponding to each feature in the data set. 
UniProt is a resource of protein sequence and functional information 
hosted by EMBL-EBI, PIR and SIB. The AlphaFold Protein Structure Database, 
developed by DeepMind and EMBL-EBI, provides open access to protein structure 
predictions for the human proteome and other key proteins of interest. 
Note that (depending on the species) many proteins are not yet covered in 
AlphaFold (in this case, the link below will lead to a non-existent page). 

```{r linktable, warning = FALSE}
linkTable <- makeDbLinkTable(
    df = as.data.frame(rowData(scePeptides)) %>%
        rownames_to_column("pid") %>%
        dplyr::select("pid", "Peptides_einprotProtein", 
                      "Peptides_einprotLabel") %>%
        dplyr::rename(einprotProtein = Peptides_einprotProtein,
                      einprotLabel = Peptides_einprotLabel), 
    idCol = "einprotProtein", 
    speciesCommon = speciesInfo$speciesCommon, 
    addSpeciesSpecificColumns = TRUE, 
    convTablePomBase = pbconv, 
    convTableWormBase = wbconv
) %>%
    dplyr::select(-"einprotProtein")
DT::datatable(
    as.data.frame(linkTable), escape = FALSE,
    filter = list(position = "top", clear = FALSE),
    options = list(scrollX = TRUE, pageLength = 20,
                   search = list(regex = FALSE, caseInsensitive = TRUE))
)
```


# Assemble SingleCellExperiment object

Finally, we assemble all the information calculated above in a 
[SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment/) 
object, which can later be used e.g. for exploration with 
[iSEE](https://bioconductor.org/packages/iSEE/) [@Rue-Albrecht2018isee].

```{r make-sce}
## Add rowData information from Protein SCE to Peptide SCE
colnames(rowData(sceProteins)) <- paste0("Proteins_", colnames(rowData(sceProteins)))
rowData(scePeptides) <- cbind(
    rowData(scePeptides),
    rowData(sceProteins)[match(rowData(scePeptides)[["Peptides_einprotMatchProt"]],
                               rowData(sceProteins)[["Proteins_einprotMatchProt"]]), ]
)

scePeptides
```

# Plot logFC of peptides vs proteins

```{r logfc-vs-logfc, warning=FALSE, fig.width=10.5, fig.height=7.5}
for (comparisonName in names(comparisonList$comparisons)) {
    if (all(c(paste0("Proteins_", comparisonName, ".logFC"),
              paste0("Peptides_", comparisonName, ".logFC"),
              paste0("PTMs_", comparisonName, ".showInVolcano")) %in% 
            colnames(rowData(scePeptides)))) {
        rescomp <- as.data.frame(rowData(scePeptides)) %>%
            rownames_to_column("pid")
        xv <- paste0("Proteins_", comparisonName, ".logFC")
        yv <- paste0("Peptides_", comparisonName, ".logFC")
        cv <- paste0("PTMs_", comparisonName, ".showInVolcano")
        print(ggplot(rescomp, 
                     aes(x = .data[[xv]],
                         y = .data[[yv]],
                         color = .data[[cv]])) + 
                  geom_abline(slope = 1, intercept = 0, linetype = "dashed") + 
                  theme_bw() + 
                  theme(axis.text = element_text(size = 12),
                        axis.title = element_text(size = 14),
                        title = element_text(size = 14),
                        legend.position = "none") +
                  labs(x = "log2(fold change), proteins", 
                       y = "log2(fold change), peptides",
                       title = comparisonName) + 
                  geom_point(fill = "lightgrey", color = "grey",
                             pch = 21, size = 1.5) +
                  geom_point(data = rescomp %>%
                                 dplyr::filter(.data[[cv]]),
                             fill = "red", color = "grey", pch = 21, size = 1.5) +
                  geom_text_repel(
                      data = rescomp %>%
                          dplyr::filter(
                              .data[[cv]] |
                                  .data$pid %in% volcanoFeaturesToLabel
                          ) %>%
                          dplyr::arrange(
                              dplyr::desc(abs(.data[[xv]] - .data[[yv]]))
                          ) %>%
                          dplyr::filter(dplyr::between(dplyr::row_number(), 0,
                                                       volcanoMaxFeatures) |
                                            .data$pid %in% volcanoFeaturesToLabel),
                      aes(label = .data$pid), max.overlaps = Inf, size = 4,
                      min.segment.length = 0.1))
    }
}
```


# Save SingleCellExperiment object

The `SingleCellExperiment` object created above is saved in the following location: 

```{r save-sce}
sceFile <- sub("\\.Rmd$", paste0("_sce.rds"), knitr::current_input(dir = TRUE))
saveRDS(scePeptides, file = sceFile)
sceFile
```

In addition, all feature information (the `rowData` of the `SingleCellExperiment`)
is written to a text file:

```{r save-rowdata}
textFile <- sub("\\.Rmd$", paste0("_feature_info.txt"),
                knitr::current_input(dir = TRUE))
write.table(as.data.frame(rowData(scePeptides)) %>%
                rownames_to_column("FeatureID"),
            file = textFile, row.names = FALSE, col.names = TRUE,
            quote = FALSE, sep = "\t")
textFile
```


# Explore the data interactively

For interactive exploration of your results, we generate a script to launch 
an adapted [iSEE](https://bioconductor.org/packages/iSEE/) interface. The 
script can be sourced from an R console: 

```{r isee-script-path, echo=FALSE}
iSEEScript <- sub("\\.Rmd$", paste0("_iSEE.R"), knitr::current_input(dir = TRUE))
```


```{r create-markdown-chunks-dynamically, include=FALSE}
out <- paste0("\n```{r, results='asis', echo=TRUE, eval=FALSE, class.source = 'fold-show'}\n\nsource('", 
              iSEEScript, "')\n\n```")
```

`r paste(knitr::knit(text = out), collapse = '\n')`


That will open up an iSEE session where you can interactively explore your data.

```{r make-isee-script, echo=FALSE}
imptd <- sub("imputed_", "", 
             grep("imputed_", SummarizedExperiment::assayNames(scePeptides), 
                  value = TRUE))
makeiSEEScript(iSEEScript = iSEEScript, sceFile = sceFile,
               aName = imptd, tests = tests,
               assayForPlots = assayForTests,
               assayForHeatmaps = "Abundance",
               includeFeatureSetTable = FALSE)
```

# Session info {#session-info}

This report was compiled with the following package versions:

<details>
<summary><b>
Click to expand
</b></summary>
```{r session-info, echo=FALSE}
sessionInfo()
```
</details>

# References


