---
title: "einprot"
output: 
    html_document:
        theme: united
        toc: true
        toc_float: true
vignette: >
  %\VignetteIndexEntry{einprot}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message = FALSE}
library(S4Vectors)
library(SummarizedExperiment)
library(einprot)
```

## Introduction

The `einprot` R package provides easy-to-use end-to-end proteomics workflows 
for data quantified with either `MaxQuant` or `Proteome Discoverer`. It was 
originally developed to support the proteomics platform at the FMI. The 
package can be installed from GitHub as follows: 

```{r, eval = FALSE}
install.packages("remotes")
remotes::install_github("fmicompbio/einprot")
```

## Running a full workflow

The main functions for running an end-to-end workflow are (depending on the 
experiment type and quantification engine)

* `runMaxQuantAnalysis()`
* `runPDTMTAnalysis()`

Each of these functions takes a set of arguments defining input files as well 
as analysis parameters, injects these into an R markdown template, and 
renders the resulting .Rmd file into an html report. In the process, a 
collection of output files are generated, including lists and plots of 
differentially abundant proteins, and a final `SingleCellExperiment` object 
for further exploration and analysis. 

For example, we will illustrate how to run the `MaxQuant` workflow using the 
example data provided by `einprot` (from the study by XXX). The workflow 
requires a `MaxQuant` proteinGroups.txt file as well as the XML parameter 
file. It also assumes that there is a column named `Gene.names` and one 
named `Majority.Protein.IDs` in the proteinGroups.txt file, holding the gene 
and protein identifiers, respectively. Thhe type of intensities to 
use for the analysis is specified via the `iColPattern` argument.
The following code runs the workflow: 

```{r, eval = FALSE}
runMaxQuantAnalysis(
    outputDir = tempdir(), 
    outputBaseName = "ExampleMaxQuantAnalysis",
    reportTitle = "MaxQuant LFQ data processing", 
    reportAuthor = "Charlotte Soneson",
    experimentInfo = list(), 
    species = "mouse",
    mqFile = system.file("extdata", "mq_example", "1356_proteinGroups.txt",
                         package = "einprot"),
    mqParameterFile = system.file("extdata", "mq_example", "1356_mqpar.xml",
                                  package = "einprot"),
    geneIdCol = "Gene.names",
    proteinIdCol = "Majority.protein.IDs",
    primaryIdType = "gene",
    iColPattern = "^LFQ\\\\.intensity\\\\.", 
    sampleAnnot = data.frame(
        sample = c("Adnp_IP04", "Adnp_IP05", "Adnp_IP06",
                   "Chd4BF_IP07", "Chd4BF_IP08", "Chd4BF_IP09",
                   "RBC_ctrl_IP01", "RBC_ctrl_IP02", "RBC_ctrl_IP03"),
        group = c("Adnp", "Adnp", "Adnp", "Chd4BF", "Chd4BF", "Chd4BF",
                  "RBC_ctrl", "RBC_ctrl", "RBC_ctrl")
    ),
    includeOnlySamples = "", 
    excludeSamples = "",
    minScore = 10, 
    minPeptides = 2, 
    imputeMethod = "MinProb",
    ctrlGroup = "RBC_ctrl", 
    allPairwiseComparisons = FALSE, 
    normMethod = "none", 
    stattest = "limma", 
    minNbrValidValues = 2,
    minlFC = 0, 
    volcanoAdjPvalThr = 0.05,
    volcanoLog2FCThr = 1, 
    volcanoMaxFeatures = 25,
    complexFDRThr = 0.1,
    seed = 42,
    includeFeatureCollections = "complexes", 
    minSizeToKeepSet = 2, 
    complexSpecies = "all", 
    complexDbPath = NULL 
)
```

This will generate a collection of output files in the designated 
output directory (here, a temporary directory). All files will be named with 
the prefix specified by `outputBaseName`. 

## Output files generated by the workflows

Running the `einprot` workflows will generate the following set of output 
files in the designated output directory: 

* A self-contained .Rmd file. In principle, this can be edited and rerun 
manually. Note, however, that if the workflow function is rerun, the file 
will be overwritten (if `forceOverwrite` argument is set to `TRUE`). 
* An html report obtained by rendering the .Rmd file.
* A `SingleCellExperiment` object serialized to an .rds file. This object 
reflects all the analysis steps of the workflow, and is exported at the end
of rendering the .Rmd file. 
* Result files from the differential expression analysis, both for individual 
proteins and for feature sets (if applicable). 
* pdf files with volcano plots for differential expression analysis, and a 
heatmap of the overall abundance matrix (only `MaxQuant` workflow).
* An R script that, when sourced, will launch an adapted `iSEE` session for 
interactive exploration of the data. 

## Getting supported species

To get the species currently supported by `einprot`, we can use the function 
`getSupportedSpecies()`. The species information is used to retrieve 
suitable feature identifiers for complexes and GO terms (if requested), and 
to provide automatic links to species-specific databases such as PomBase 
and WormBase, whenever applicable. 

```{r}
getSupportedSpecies()
```

## Listing available complex DBs

If specified by the user, `einprot` will test for differential abundance not 
only of individual proteins, but also of feature sets. Currently, GO terms 
and protein complexes (plus any user-specified sets) are supported. The 
built-in complex database was created using the `makeComplexDB()` function, 
and combines annotated complexes for multiple different species. Regardless 
of where it was originally defined, each complex is also lifted over to the 
current species of interest, using the ortholog conversion functionality of 
the `babelgene` package. To list the available complex databases provided 
by `einprot`, we can use the `listComplexDBs()` function: 

```{r}
## List available complex databases
(lcdb <- listComplexDBs())

## Read the most recent one
cdb <- readRDS(lcdb$complexDbPath[1])

## cdb is a list of databases - each one containing identifiers for a 
## specific species
class(cdb)
names(cdb)

cdb$mouse

## Additional information about the complexes
mcols(cdb$mouse)
```

## Checking the intensity column pattern

To check which columns will be extracted from the input text file to generate
the main intensity matrix, we can use the `getIntensityColumns()` function, 
provided with the path to the input text file and the desired `iColPattern`.

```{r}
mqFile <- system.file("extdata", "mq_example", "1356_proteinGroups.txt",
                      package = "einprot")
```

```{r}
getIntensityColumns(inFile = mqFile, iColPattern = "^LFQ\\.intensity\\.")
```

## Reading data into a `SingleCellExperiment` object

`einprot` has functionality for reading data from `MaxQuant` and 
`ProteomeDiscoverer` protein quantification text files and generating a 
`SingleCellExperiment` object. All columns in the text file that are 
typically sample-specific will be converted into assays, while all other 
columns will be added as row annotations. 

```{r}
imp <- importExperiment(mqFile, iColPattern = "^LFQ\\.intensity\\.")

## The aName output provides the name of the "main" assay 
## (corresponding to the specified iColPattern)
imp$aName

## Imported SingleCellExperiment object
imp$sce

## All assays
assayNames(imp$sce)

## Row annotation
rowData(imp$sce)
```

## Querying the `MaxQuant`/`Proteome Discoverer` metadata

In addition to reading the protein quantifications, `einprot` also reads the 
metadata XML files generated by `MaxQuant` and `Proteome Discoverer` and 
extracts information to display in the final report. We can also convert it 
to a nice-looking table. The `Proteome Discoverer` summary function 
requires the path to a folder that contains at least the `_InputFiles.txt` 
and `_StudyInformation.txt` output files from the `Proteome Discoverer` run, 
in addition to the `.pdAnalysis` file with the run metadata. 

```{r}
## MaxQuant
mqinfo <- readMaxQuantXML(
    system.file("extdata", "mq_example", "1356_mqpar.xml", package = "einprot"))
makeTableFromList(mqinfo)
```

```{r}
## Proteome Discoverer
pdinfo <- readProteomeDiscovererInfo(
    pdOutputFolder = system.file("extdata", "pdtmt_example", 
                                 package = "einprot"),
    pdResultName = "Fig2_m23139_RTS_QC_varMods", 
    pdAnalysisFile = system.file("extdata", "pdtmt_example", 
                                 "Fig2_m23139_RTS_QC_varMods.pdAnalysis",
                                 package = "einprot"))
makeTableFromList(pdinfo)
```


## Session info

<details>
<summary><b>
Click to expand
</b></summary>
```{r session-info, echo=FALSE}
sessionInfo()
```
</details>

