---
title: 'einprot: flexible, easy-to-use, reproducible workflows for statistical analysis of quantitative proteomics data'
tags:
  - R
  - proteomics
authors:
  - name: Charlotte Soneson
    orcid: 0000-0003-3833-2169
    affiliation: "1, 2"
    corresponding: true
  - name: Vytautas Iesmantavicius
    orcid: 0000-0002-2512-9957
    affiliation: 1
  - name: Daniel Hess
    orcid: 0000-0002-1642-5404
    affiliation: 1
  - name: Michael B Stadler
    orcid: 0000-0002-2269-4934
    affiliation: "1, 2, 3"
  - name: Jan Seebacher
    orcid: 0000-0002-7858-2720
    affiliation: 1
affiliations:
 - name: Friedrich Miescher Institute for Biomedical Research, Basel, Switzerland
   index: 1
 - name: SIB Swiss Institute of Bioinformatics, Basel, Switzerland
   index: 2
 - name: University of Basel, Switzerland
   index: 3
date: 29 July 2023
bibliography: einprot.bib
---

# Summary

Quantitative proteomics has come a long way - what used to be specialized analyses performed in proteomics research groups is nowadays a routine service in many proteomics core facilities, and a large collection of sophisticated quantification and analysis tools are available. 
Yet, the necessary reporting tasks, including statistical analyses of the resulting data, as well as describing all data processing steps, providing quality control, exploration opportunities and result visualizations for publication in a user-friendly way, are generally not routine or automated, and many different analysis workflows are conceivable [@Peng2023-optim]. 
Moreover, additional downstream analyses and integration with other types of data are often necessary, and these are more likely to succeed when all steps of the routine data analysis workflow are transparent and well documented. 

The einprot R package provides accessible workflows accommodating quality control, filtering, exploratory analysis and statistical analysis of proteomics data quantified with several commonly used tools, including label-free quantification (LFQ) with MaxQuant [@Cox2008-mq] or FragPipe [@Kong2017-msfragger], and tandem mass tag (TMT)-multiplexed data quantified with Proteome Discoverer [@Orsburn2021-pd]. 
Each workflow is provided in the form of a template R Markdown (Rmd) file [@Xie2018-rmarkdown], containing the code to be executed as well as text descriptions and explanations of the individual steps. 
This can significantly reduce the amount of time spent on routine processing tasks, e.g., for a core facility, and enables the entire analysis process to be shared with collaborators or data generators in a way that is comprehensive and easy to follow. 
In addition, the report contains several interactive tables and plots that allow users to immediately explore their data. 

To run a workflow, the user calls a single function in their R session, to which they provide the path to the quantification file(s) as well as a number of additional arguments specifying details about the experiment and the requested analyses. 
einprot then copies the template Rmd script from the package location to the designated output directory, injects the arguments specified by the user to create a stand-alone file, and compiles this into an html report describing the complete analysis. 
The stand-alone Rmd file is retained and can, if necessary, be manually modified and fine-tuned by the user and recompiled. 
Alternatively, the analyst can provide their own template Rmd file if custom analyses are desired. 
A collection of example reports generated by einprot are provided at [https://csoneson.github.io/einprot_examples/](https://csoneson.github.io/einprot_examples/). 
The einprot functions used in the workflows can also be called directly in the R session for an interactive analysis or recreation of specific plots (see the online [vignette](https://fmicompbio.github.io/einprot/articles/einprot.html) for examples). 
While compiling the report, einprot exports a set of text files and publication-ready plots for further inspection and dissemination of the results (see the [vignette](https://fmicompbio.github.io/einprot/articles/einprot.html) for a full list of output files and \autoref{fig:1} and \autoref{fig:2} for examples of figures generated by the workflows).

![Example figures generated by the einprot workflow, based on a data set from @Nie2021-covid, requantified by @He2022-covid. The figures can also be generated programmatically from the SingleCellExperiment object generated by the einprot workflow. A. Distribution of abundances in each sample. B. Principal component analysis representation of samples, percent variance explained by each principal component, and the proteins with the highest loadings in the first two components. C. Histograms showing the distribution of observed and imputed abundance values in each sample. D. Volcano plot for one of the tested contrasts. E. MA plot for the contrast shown in D. \label{fig:1}](einprot_fig1.png)

![Example figures generated by the einprot workflow, based on an IP-MS data set from @Ostapcuk2018-adnp, quantified with MaxQuant. A. Overview of the proteins filtered out by the workflow. B. Distribution of LFQ intensities across samples. C. Global missing value pattern. D. Volcano plot for one of the tested contrasts. E. List of known protein complexes most strongly associated with the contrast in D. F. The same volcano plot as in D, but with the members of the top-ranked complex from E highlighted. \label{fig:2}](einprot_fig2.png)

The current version of einprot fully supports five common organisms: _Mus musculus_, _Homo sapiens_, _Caenorhabditis elegans_, _Saccharomyces cerevisiae_ and _Schizosaccharomyces pombe_. 
For these organisms, the user can elect to perform automatic enrichment testing of Gene Ontology terms [@Ashburner2000-go; @Gene_Ontology_Consortium2023-go] and known protein complexes, obtained from CORUM [@Giurgiu2019-corum], PomBase [@Harris2022-pombase] and CYC2008 [@Pu2009-cyc2008] and mapped to the organism of interest using the ortholog mapping from the babelgene R package [@Dolgalev2022-babelgene]. 
Other species can be analyzed by skipping (parts of) the automatic enrichment testing. 
Finally, for improved interoperability with other tools, especially from the Bioconductor ecosystem [@Huber2015-bioc], einprot stores all raw and processed values (including, e.g., results from differential abundance analysis and dimensionality reduction via principal component analysis) in a SingleCellExperiment object [@Amezquita2020-sce]. 
In addition, einprot automatically generates an R script that can be sourced to launch a customized interactive exploration session using the iSEE R package [@Rue-Albrecht2018-isee] (\autoref{fig:3}).

![A part of the interactive interface configured for each einprot run, based on the iSEE package. A variety of interactive panels allows the user to explore the statistical results, abundance values and imputation status for individual proteins, a low-dimensional representation of the samples, as well as to browse all results in table form. In addition, all the flexibility provided by the iSEE package is retained, and the user can shape the interface and panel composition according to their own needs. \label{fig:3}](einprot_fig3.png)

# Statement of need

Several other toolkits are available for analyzing proteomics data [@Bai2023-benchmark]. 
Arguably, routine statistical analyses are most commonly performed using vendor- or community-developed software suites accessible via graphical user interfaces, such as Perseus [@Tyanova2016-perseus] or Proteome Discoverer [@Orsburn2021-pd].
While they provide comprehensive analysis capabilities, they are typically not fully open source, and sometimes only available for selected operating systems. 
In addition, the manual "point-and-click" interface means that they are difficult to incorporate into an automated or programmable data analysis pipeline. 
Among the broad range of community-developed open-source tools, many (e.g., POMAShiny [@Castellano-Escuder2021-pomashiny], LFQ-Analyst [@Shah2020-lfqanalyst], amica [@Didusch2022-amica], Eatomics [@Kraus2021-eatomics], ProVision [@Gallant2020-provision], and the proteomics workflows provided via the Galaxy platform [@Galaxy_Community2022-galaxy; @Hiltemann2023-gtn]) provide a graphical user interface where the user steps through the analysis manually; some additionally allow a summary report to be exported. 
Other software packages provide extensive collections of analysis functions, but require familiarity with programming to use (e.g. protti [@Quast2021-protti], POMA [@Castellano-Escuder2021-pomashiny], SafeQuant [@Glatter2012-safequant], Proteus [@Gierlinski2018-proteus], DEP [@Zhang2018-dep], prolfqua [@Wolski2023-prolfqua], MSstats [@Choi2014-msstats]). 
With einprot, we attempt to hit a middle ground by providing a collection of fully reproducible workflows that do not require extensive coding skills to run, yet return comprehensive, self-contained reports that contain all the executed code and are fully customizable if needed. 
In addition, by returning the data and results in a standardized format, the interoperability with other packages, including some of the ones mentioned above, is simplified and allows the user to take advantage of a vast ecosystem of tools. 
One such example is the direct interface to iSEE, which lets the user interactively explore all the results and processed values generated by and documented in the reproducible workflows. 
In this way, einprot is used productively by the proteomics and protein analysis facility at FMI and as the basis for published and submitted articles [@Welte2023-conv]. 

# Availability and installation

einprot is available from GitHub ([https://github.com/fmicompbio/einprot](https://github.com/fmicompbio/einprot)) and can be installed using the standard installation processes for R packages (e.g., using the remotes package [@Csardi2021-remotes]). 
Documentation and example usage are available from [https://fmicompbio.github.io/einprot/](https://fmicompbio.github.io/einprot/). 
A collection of example reports can be browsed at [https://csoneson.github.io/einprot_examples/](https://csoneson.github.io/einprot_examples/). 

# Acknowledgements

The authors would like to thank Merle Skribbe, Seraina Steiger, Thomas Welte, Patrick Matthias, Helge Grosshans and Marc BÃ¼hler for testing and feedback on the software, and Laurent Gatto for valuable discussions. 
This work was supported by the Novartis Research Foundation. 
The funding body did not have any role in the design of the study, the collection, analysis, and interpretation of data, or in writing the manuscript. 

# References
