<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>einprot • einprot</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="einprot">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">einprot</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/einprot.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="einprot_files/kePrint-0.0.1/kePrint.js"></script><link href="einprot_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>einprot</h1>
            
            <h4 data-toc-skip class="date">2025-03-15</h4>
      

      <div class="hidden name"><code>einprot.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/S4Vectors" class="external-link">S4Vectors</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">einprot</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>einprot</code> R package provides easy-to-use end-to-end
proteomics workflows for data quantified with either
<code>MaxQuant</code> <span class="citation">(Cox and Mann 2008)</span>
(LFQ), <code>Proteome Discoverer</code> <span class="citation">(Orsburn
2021)</span> (TMT), or <a href="https://fragpipe.nesvilab.org/" class="external-link"><code>FragPipe</code></a>. It was
originally developed to support the proteomics platform at the FMI. The
package can be installed from GitHub as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"fmicompbio/einprot"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-a-full-workflow">Running a full workflow<a class="anchor" aria-label="anchor" href="#running-a-full-workflow"></a>
</h2>
<p>The main functions for running an end-to-end workflow are (depending
on the experiment type and quantification engine)</p>
<ul>
<li><code><a href="../reference/runMaxQuantAnalysis.html">runMaxQuantAnalysis()</a></code></li>
<li><code><a href="../reference/runPDTMTAnalysis.html">runPDTMTAnalysis()</a></code></li>
<li><code><a href="../reference/runPDTMTptmAnalysis.html">runPDTMTptmAnalysis()</a></code></li>
<li><code><a href="../reference/runFragPipeAnalysis.html">runFragPipeAnalysis()</a></code></li>
</ul>
<p>Each of these functions takes a set of arguments defining input files
as well as analysis parameters, injects these into an R markdown
template, and renders the resulting .Rmd file into an html report. In
the process, a collection of output files are generated, including lists
and plots of differentially abundant proteins, and a final
<code>SingleCellExperiment</code> object for further exploration and
analysis.</p>
<p>As an example, we will illustrate how to run the
<code>MaxQuant</code> workflow using the example data provided in
<code>einprot</code> (from <span class="citation">Ostapcuk et al.
(2018)</span>). The process is similar for the other workflows. The
<code>MaxQuant</code> workflow requires a <code>MaxQuant</code>
<code>proteinGroups.txt</code> file as well as (optionally) the XML
parameter file. The type of intensities to use for the analysis is
specified via the <code>iColPattern</code> argument. The following code
runs the workflow, using mostly default parameter values:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sampleAnnot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mq_example"</span>, <span class="st">"1356_sampleAnnot.txt"</span>,</span>
<span>                                       package <span class="op">=</span> <span class="st">"einprot"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMaxQuantAnalysis.html">runMaxQuantAnalysis</a></span><span class="op">(</span></span>
<span>    outputDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>    outputBaseName <span class="op">=</span> <span class="st">"MQ_LFQ"</span>,</span>
<span>    reportTitle <span class="op">=</span> <span class="st">"MaxQuant LFQ data processing"</span>, </span>
<span>    reportAuthor <span class="op">=</span> <span class="st">"Charlotte Soneson"</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"mouse"</span>,</span>
<span>    mqFile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mq_example"</span>, <span class="st">"1356_proteinGroups.txt"</span>,</span>
<span>                         package <span class="op">=</span> <span class="st">"einprot"</span><span class="op">)</span>,</span>
<span>    mqParameterFile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mq_example"</span>, <span class="st">"1356_mqpar.xml"</span>,</span>
<span>                                  package <span class="op">=</span> <span class="st">"einprot"</span><span class="op">)</span>,</span>
<span>    iColPattern <span class="op">=</span> <span class="st">"^LFQ\\.intensity\\."</span>, </span>
<span>    sampleAnnot <span class="op">=</span> <span class="va">sampleAnnot</span>,</span>
<span>    imputeMethod <span class="op">=</span> <span class="st">"MinProb"</span>,</span>
<span>    ctrlGroup <span class="op">=</span> <span class="st">"RBC_ctrl"</span>, </span>
<span>    allPairwiseComparisons <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    normMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>    stattest <span class="op">=</span> <span class="st">"limma"</span>, </span>
<span>    includeFeatureCollections <span class="op">=</span> <span class="st">"complexes"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This run will generate a collection of output files in the designated
output directory (here, a temporary directory). All files will be named
with the prefix specified by <code>outputBaseName</code>.</p>
</div>
<div class="section level2">
<h2 id="output-files-generated-by-the-workflows">Output files generated by the workflows<a class="anchor" aria-label="anchor" href="#output-files-generated-by-the-workflows"></a>
</h2>
<p>Running the <code>einprot</code> workflow above will generate the
following set of output files in the designated output directory:</p>
<p><img src="einprot_output_list.png"></p>
<ul>
<li>A self-contained .Rmd file (here, <code>MQ_LFQ.Rmd</code>) with all
the code run in the workflow. In principle, this can be edited and rerun
manually. Note, however, that if the workflow function is called again,
the file will be overwritten (if <code>forceOverwrite</code> argument is
set to <code>TRUE</code>).</li>
<li>An html report (<code>MQ_LFQ.html</code>) obtained by rendering the
.Rmd file.</li>
<li>A <code>SingleCellExperiment</code> object serialized to an .rds
file (<code>MQ_LFQ_sce.rds</code>). This object reflects all the
analysis steps of the workflow, and is exported at the end of rendering
the .Rmd file. Very long annotation columns are stored in a separate
file (<code>MQ_LFQ_sce_extra_annots.tsv</code>).</li>
<li>Result files from the differential expression analysis, both for
individual proteins and for feature sets (known protein complexes and/or
GO terms, if applicable) (<code>MQ_LFQ_testres_*.txt</code>), and
exported abundance values for all significant features
(<code>MQ_LFQ_abundance_values_significant.txt</code>).</li>
<li>pdf files with volcano plots for differential abundance analysis
(<code>MQ_LFQ_volcano_*.pdf</code>), and a heatmap of the overall
abundance matrix (<code>MQ_LFQ_heatmap_centered.pdf</code>, only for the
<code>MaxQuant</code> and <code>FragPipe</code> workflows).</li>
<li>An R script that, when sourced, will launch an adapted
<code>iSEE</code> <span class="citation">(Rue-Albrecht et al.
2018)</span> session for interactive exploration of the data
(<code>MQ_LFQ_iSEE.R</code>, which incorporates an interactive guided
tour specified in <code>MQ_LFQ_iSEE_tour.csv</code>).</li>
<li>Additional information about retained and filtered-out features
(<code>MQ_LFQ_feature_info.txt</code>,
<code>MQ_LFQ_filtered_out_features.txt</code>).</li>
</ul>
<p>All file names start with the <code>outputBaseName</code> specified
in the call to <code>runMaxQuantAnalysis</code> (here,
<code>MQ_LFQ</code>).</p>
</div>
<div class="section level2">
<h2 id="listing-supported-species">Listing supported species<a class="anchor" aria-label="anchor" href="#listing-supported-species"></a>
</h2>
<p>The list of species currently supported by <code>einprot</code> can
be obtained using the function <code><a href="../reference/getSupportedSpecies.html">getSupportedSpecies()</a></code>. The
species information is used to retrieve suitable feature identifiers for
complexes and GO terms (if requested), and to provide automatic links to
species-specific databases such as PomBase and WormBase, whenever
applicable.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSupportedSpecies.html">getSupportedSpecies</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;     taxId                         species        speciesCommon</span></span>
<span><span class="co">#&gt; 1   10090                    Mus musculus                mouse</span></span>
<span><span class="co">#&gt; 2    9606                    Homo sapiens                human</span></span>
<span><span class="co">#&gt; 3    6239          Caenorhabditis elegans            roundworm</span></span>
<span><span class="co">#&gt; 4    7955                     Danio rerio            zebrafish</span></span>
<span><span class="co">#&gt; 5    7227         Drosophila melanogaster             fruitfly</span></span>
<span><span class="co">#&gt; 6    4932        Saccharomyces cerevisiae        baker's yeast</span></span>
<span><span class="co">#&gt; 7  284812 Schizosaccharomyces pombe 972h-        fission yeast</span></span>
<span><span class="co">#&gt; 8   28377             Anolis carolinensis          green anole</span></span>
<span><span class="co">#&gt; 9    9913                      Bos taurus               bovine</span></span>
<span><span class="co">#&gt; 10   9615          Canis lupus familiaris                  dog</span></span>
<span><span class="co">#&gt; 11   9796                  Equus caballus                horse</span></span>
<span><span class="co">#&gt; 12   9685                     Felis catus                  cat</span></span>
<span><span class="co">#&gt; 13   9031                   Gallus gallus              chicken</span></span>
<span><span class="co">#&gt; 14   9544                  Macaca mulatta       rhesus macaque</span></span>
<span><span class="co">#&gt; 15  13616           Monodelphis domestica              opossum</span></span>
<span><span class="co">#&gt; 16   9258        Ornithorhynchus anatinus             platypus</span></span>
<span><span class="co">#&gt; 17   9598                 Pan troglodytes           chimpanzee</span></span>
<span><span class="co">#&gt; 18  10116               Rattus norvegicus           Norway rat</span></span>
<span><span class="co">#&gt; 19   9823                      Sus scrofa                  pig</span></span>
<span><span class="co">#&gt; 20   8364              Xenopus tropicalis tropical clawed frog</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="listing-available-complex-dbs">Listing available complex DBs<a class="anchor" aria-label="anchor" href="#listing-available-complex-dbs"></a>
</h2>
<p>If specified by the user, <code>einprot</code> will test for
differential abundance not only of individual proteins, but also of
feature sets. Currently, GO terms and protein complexes (plus any
user-specified sets) are supported. The built-in complex database was
created using the <code><a href="../reference/makeComplexDB.html">makeComplexDB()</a></code> function, and combines
annotated complexes for multiple different species. Regardless of where
it was originally defined, each complex is also lifted over to the
current species of interest, using the ortholog conversion functionality
of the <code>babelgene</code> package. To list the available complex
databases provided by <code>einprot</code>, we can use the
<code><a href="../reference/listComplexDBs.html">listComplexDBs()</a></code> function:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## List available complex databases</span></span>
<span><span class="op">(</span><span class="va">lcdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/listComplexDBs.html">listComplexDBs</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                                              complexDbPath</span></span>
<span><span class="co">#&gt; 1 /Users/runner/work/_temp/Library/einprot/extdata/complexes/complexdb_einprot0.9.3_20240328_orthologs.rds</span></span>
<span><span class="co">#&gt; 2 /Users/runner/work/_temp/Library/einprot/extdata/complexes/complexdb_einprot0.5.0_20220323_orthologs.rds</span></span>
<span><span class="co">#&gt;      genDate</span></span>
<span><span class="co">#&gt; 1 2024-03-28</span></span>
<span><span class="co">#&gt; 2 2022-03-23</span></span>
<span></span>
<span><span class="co">## Read the most recent one</span></span>
<span><span class="va">cdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">lcdb</span><span class="op">$</span><span class="va">complexDbPath</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## cdb is a list of databases - each one containing identifiers for a </span></span>
<span><span class="co">## specific species</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">cdb</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "list"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cdb</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "mouse"                           "human"                          </span></span>
<span><span class="co">#&gt; [3] "baker's yeast"                   "Caenorhabditis elegans"         </span></span>
<span><span class="co">#&gt; [5] "Schizosaccharomyces pombe 972h-"</span></span>
<span></span>
<span><span class="va">cdb</span><span class="op">$</span><span class="va">mouse</span></span>
<span><span class="co">#&gt; CharacterList of length 11257</span></span>
<span><span class="co">#&gt; [["mouse: (ER)-localized multiprotein complex, Ig heavy chains associated"]] ...</span></span>
<span><span class="co">#&gt; [["mouse: (ER)-localized multiprotein complex, in absence of Ig heavy chains"]] </span></span>
<span><span class="co">#&gt; [["mouse: 20S proteasome"]] Psmb1 Psmb5 Psma3 Psma2 ... Psmb2 Psma1 Psma7 Psma5</span></span>
<span><span class="co">#&gt; [["mouse: A2m-Anxa6-Lrp1 complex"]] A2M ANXA6 Anxa6 LRP1 A2m Lrp1</span></span>
<span><span class="co">#&gt; [["mouse: Abcg5-Abcg1 complex"]] Abcg1 Abcg5</span></span>
<span><span class="co">#&gt; [["mouse: Abcg5-Abcg2 complex"]] Abcg2 Abcg5</span></span>
<span><span class="co">#&gt; [["mouse: Abcg5-Abcg4 complex"]] Abcg4 Abcg5</span></span>
<span><span class="co">#&gt; [["mouse: Abcg5-Abcg8 complex (+1 alt. ID)"]] Abcg5 Abcg8</span></span>
<span><span class="co">#&gt; [["mouse: Abcg8-Abcg4 complex"]] Abcg4 Abcg8</span></span>
<span><span class="co">#&gt; [["mouse: Abi1-Wasl complex (+1 alt. ID)"]] Abi1 Wasl</span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; &lt;11247 more elements&gt;</span></span>
<span></span>
<span><span class="co">## Additional information about the complexes</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">cdb</span><span class="op">$</span><span class="va">mouse</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 11257 rows and 4 columns</span></span>
<span><span class="co">#&gt;                                                                           Species.common</span></span>
<span><span class="co">#&gt;                                                                              &lt;character&gt;</span></span>
<span><span class="co">#&gt; mouse: (ER)-localized multiprotein complex, Ig heavy chains associated             mouse</span></span>
<span><span class="co">#&gt; mouse: (ER)-localized multiprotein complex, in absence of Ig heavy chains          mouse</span></span>
<span><span class="co">#&gt; mouse: 20S proteasome                                                              mouse</span></span>
<span><span class="co">#&gt; mouse: A2m-Anxa6-Lrp1 complex                                                      mouse</span></span>
<span><span class="co">#&gt; mouse: Abcg5-Abcg1 complex                                                         mouse</span></span>
<span><span class="co">#&gt; ...                                                                                  ...</span></span>
<span><span class="co">#&gt; S.pombe: vacuolar proton-transporting V-type ATPase, V1 domain             Schizosacc...</span></span>
<span><span class="co">#&gt; S.pombe: voltage-gated potassium channel complex                           Schizosacc...</span></span>
<span><span class="co">#&gt; S.pombe: Vps55/Vps68 complex                                               Schizosacc...</span></span>
<span><span class="co">#&gt; S.pombe: Wpl/Pds5 cohesin loading/unloading complex                        Schizosacc...</span></span>
<span><span class="co">#&gt; S.pombe: zeta DNA polymerase complex                                       Schizosacc...</span></span>
<span><span class="co">#&gt;                                                                                Source</span></span>
<span><span class="co">#&gt;                                                                           &lt;character&gt;</span></span>
<span><span class="co">#&gt; mouse: (ER)-localized multiprotein complex, Ig heavy chains associated          CORUM</span></span>
<span><span class="co">#&gt; mouse: (ER)-localized multiprotein complex, in absence of Ig heavy chains       CORUM</span></span>
<span><span class="co">#&gt; mouse: 20S proteasome                                                           CORUM</span></span>
<span><span class="co">#&gt; mouse: A2m-Anxa6-Lrp1 complex                                                   CORUM</span></span>
<span><span class="co">#&gt; mouse: Abcg5-Abcg1 complex                                                      CORUM</span></span>
<span><span class="co">#&gt; ...                                                                               ...</span></span>
<span><span class="co">#&gt; S.pombe: vacuolar proton-transporting V-type ATPase, V1 domain                pombase</span></span>
<span><span class="co">#&gt; S.pombe: voltage-gated potassium channel complex                              pombase</span></span>
<span><span class="co">#&gt; S.pombe: Vps55/Vps68 complex                                                  pombase</span></span>
<span><span class="co">#&gt; S.pombe: Wpl/Pds5 cohesin loading/unloading complex                           pombase</span></span>
<span><span class="co">#&gt; S.pombe: zeta DNA polymerase complex                                          pombase</span></span>
<span><span class="co">#&gt;                                                                                    PMID</span></span>
<span><span class="co">#&gt;                                                                             &lt;character&gt;</span></span>
<span><span class="co">#&gt; mouse: (ER)-localized multiprotein complex, Ig heavy chains associated         12475965</span></span>
<span><span class="co">#&gt; mouse: (ER)-localized multiprotein complex, in absence of Ig heavy chains      12475965</span></span>
<span><span class="co">#&gt; mouse: 20S proteasome                                                          10436176</span></span>
<span><span class="co">#&gt; mouse: A2m-Anxa6-Lrp1 complex                                                  15226301</span></span>
<span><span class="co">#&gt; mouse: Abcg5-Abcg1 complex                                                     14504269</span></span>
<span><span class="co">#&gt; ...                                                                                 ...</span></span>
<span><span class="co">#&gt; S.pombe: vacuolar proton-transporting V-type ATPase, V1 domain            14735354;G...</span></span>
<span><span class="co">#&gt; S.pombe: voltage-gated potassium channel complex                          GO_REF:000...</span></span>
<span><span class="co">#&gt; S.pombe: Vps55/Vps68 complex                                              GO_REF:000...</span></span>
<span><span class="co">#&gt; S.pombe: Wpl/Pds5 cohesin loading/unloading complex                            26687354</span></span>
<span><span class="co">#&gt; S.pombe: zeta DNA polymerase complex                                      GO_REF:000...</span></span>
<span><span class="co">#&gt;                                                                               All.names</span></span>
<span><span class="co">#&gt;                                                                             &lt;character&gt;</span></span>
<span><span class="co">#&gt; mouse: (ER)-localized multiprotein complex, Ig heavy chains associated    mouse: (ER...</span></span>
<span><span class="co">#&gt; mouse: (ER)-localized multiprotein complex, in absence of Ig heavy chains mouse: (ER...</span></span>
<span><span class="co">#&gt; mouse: 20S proteasome                                                     mouse: 20S...</span></span>
<span><span class="co">#&gt; mouse: A2m-Anxa6-Lrp1 complex                                             mouse: A2m...</span></span>
<span><span class="co">#&gt; mouse: Abcg5-Abcg1 complex                                                mouse: Abc...</span></span>
<span><span class="co">#&gt; ...                                                                                 ...</span></span>
<span><span class="co">#&gt; S.pombe: vacuolar proton-transporting V-type ATPase, V1 domain            S.pombe: v...</span></span>
<span><span class="co">#&gt; S.pombe: voltage-gated potassium channel complex                          S.pombe: v...</span></span>
<span><span class="co">#&gt; S.pombe: Vps55/Vps68 complex                                              S.pombe: V...</span></span>
<span><span class="co">#&gt; S.pombe: Wpl/Pds5 cohesin loading/unloading complex                       S.pombe: W...</span></span>
<span><span class="co">#&gt; S.pombe: zeta DNA polymerase complex                                      S.pombe: z...</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="checking-the-intensity-column-pattern">Checking the intensity column pattern<a class="anchor" aria-label="anchor" href="#checking-the-intensity-column-pattern"></a>
</h2>
<p>As described above, the workflow functions require the specification
of a ‘column pattern’, that determines which columns will be extracted
from the input text file to generate the main intensity matrix. To check
which columns will be extracted with a given column pattern, we can use
the <code><a href="../reference/getIntensityColumns.html">getIntensityColumns()</a></code> function, provided with the path
to the input text file and the desired <code>iColPattern</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mqFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mq_example"</span>, <span class="st">"1356_proteinGroups.txt"</span>,</span>
<span>                      package <span class="op">=</span> <span class="st">"einprot"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getIntensityColumns.html">getIntensityColumns</a></span><span class="op">(</span>inFile <span class="op">=</span> <span class="va">mqFile</span>, iColPattern <span class="op">=</span> <span class="st">"^LFQ\\.intensity\\."</span><span class="op">)</span></span>
<span><span class="co">#&gt; $iColsAll</span></span>
<span><span class="co">#&gt; [1] "LFQ.intensity.Adnp_IP04"     "LFQ.intensity.Adnp_IP05"    </span></span>
<span><span class="co">#&gt; [3] "LFQ.intensity.Adnp_IP06"     "LFQ.intensity.Chd4BF_IP07"  </span></span>
<span><span class="co">#&gt; [5] "LFQ.intensity.Chd4BF_IP08"   "LFQ.intensity.Chd4BF_IP09"  </span></span>
<span><span class="co">#&gt; [7] "LFQ.intensity.RBC_ctrl_IP01" "LFQ.intensity.RBC_ctrl_IP02"</span></span>
<span><span class="co">#&gt; [9] "LFQ.intensity.RBC_ctrl_IP03"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $iCols</span></span>
<span><span class="co">#&gt; [1] "LFQ.intensity.Adnp_IP04"     "LFQ.intensity.Adnp_IP05"    </span></span>
<span><span class="co">#&gt; [3] "LFQ.intensity.Adnp_IP06"     "LFQ.intensity.Chd4BF_IP07"  </span></span>
<span><span class="co">#&gt; [5] "LFQ.intensity.Chd4BF_IP08"   "LFQ.intensity.Chd4BF_IP09"  </span></span>
<span><span class="co">#&gt; [7] "LFQ.intensity.RBC_ctrl_IP01" "LFQ.intensity.RBC_ctrl_IP02"</span></span>
<span><span class="co">#&gt; [9] "LFQ.intensity.RBC_ctrl_IP03"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reading-data-into-a-singlecellexperiment-object">Reading data into a <code>SingleCellExperiment</code> object<a class="anchor" aria-label="anchor" href="#reading-data-into-a-singlecellexperiment-object"></a>
</h2>
<p><code>einprot</code> contains functionality for reading data from
<code>MaxQuant</code>, <code>FragPipe</code> and
<code>ProteomeDiscoverer</code> protein quantification text files and
generating a <code>SingleCellExperiment</code> object. All columns in
the text file that are typically sample-specific will be converted into
assays, while all other columns will be added as row annotations.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExperiment.html">importExperiment</a></span><span class="op">(</span><span class="va">mqFile</span>, iColPattern <span class="op">=</span> <span class="st">"^LFQ\\.intensity\\."</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## The aName output provides the name of the "main" assay </span></span>
<span><span class="co">## (corresponding to the specified iColPattern)</span></span>
<span><span class="va">imp</span><span class="op">$</span><span class="va">aName</span></span>
<span><span class="co">#&gt; [1] "LFQ.intensity"</span></span>
<span></span>
<span><span class="co">## Imported SingleCellExperiment object</span></span>
<span><span class="va">imp</span><span class="op">$</span><span class="va">sce</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 463 9 </span></span>
<span><span class="co">#&gt; metadata(1): colList</span></span>
<span><span class="co">#&gt; assays(9): LFQ.intensity MS.MS.Count ... iBAQ Identification.type</span></span>
<span><span class="co">#&gt; rownames(463): 1 2 ... 462 463</span></span>
<span><span class="co">#&gt; rowData names(35): Protein.IDs Majority.protein.IDs ...</span></span>
<span><span class="co">#&gt;   Oxidation.M.site.IDs Oxidation.M.site.positions</span></span>
<span><span class="co">#&gt; colnames(9): Adnp_IP04 Adnp_IP05 ... RBC_ctrl_IP02 RBC_ctrl_IP03</span></span>
<span><span class="co">#&gt; colData names(0):</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span></span>
<span><span class="co">## All assays</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">imp</span><span class="op">$</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "LFQ.intensity"         "MS.MS.Count"           "Intensity"            </span></span>
<span><span class="co">#&gt; [4] "Sequence.coverage"     "Unique.peptides"       "Razor.unique.peptides"</span></span>
<span><span class="co">#&gt; [7] "Peptides"              "iBAQ"                  "Identification.type"</span></span>
<span></span>
<span><span class="co">## Row annotation</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">imp</span><span class="op">$</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 463 rows and 35 columns</span></span>
<span><span class="co">#&gt;       Protein.IDs Majority.protein.IDs Peptide.counts.all</span></span>
<span><span class="co">#&gt;       &lt;character&gt;          &lt;character&gt;        &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1   A0A023T672...        A0A023T672...              1;1;1</span></span>
<span><span class="co">#&gt; 2   A0A087WPL5...        A0A087WPL5...      13;13;11;1...</span></span>
<span><span class="co">#&gt; 3   Q3UH28;Q3U...        Q3UH28;Q3U...      3;3;3;3;3;...</span></span>
<span><span class="co">#&gt; 4   A2A513;CON...        A2A513;CON...      14;14;14;1...</span></span>
<span><span class="co">#&gt; 5   A2A791;A2A...        A2A791;A2A...           56;55;44</span></span>
<span><span class="co">#&gt; ...           ...                  ...                ...</span></span>
<span><span class="co">#&gt; 459 REV__P2873...        REV__P2873...                  1</span></span>
<span><span class="co">#&gt; 460 REV__P5871...        REV__P5871...                  1</span></span>
<span><span class="co">#&gt; 461 REV__Q8BHW...        REV__Q8BHW...                  1</span></span>
<span><span class="co">#&gt; 462 V9GWW1;Q68...        V9GWW1;Q68...      42;40;27;2...</span></span>
<span><span class="co">#&gt; 463        V9GXP2               V9GXP2                 29</span></span>
<span><span class="co">#&gt;     Peptide.counts.razor.unique Peptide.counts.unique Protein.names</span></span>
<span><span class="co">#&gt;                     &lt;character&gt;           &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1                         1;1;1                 1;1;1 RNA-bindin...</span></span>
<span><span class="co">#&gt; 2                 13;13;11;1...         13;13;11;1... ATP-depend...</span></span>
<span><span class="co">#&gt; 3                 3;3;3;3;3;...         3;3;3;3;3;...              </span></span>
<span><span class="co">#&gt; 4                     1;1;1;1;1             1;1;1;1;1 Keratin, t...</span></span>
<span><span class="co">#&gt; 5                      56;55;44              56;55;44 Zinc finge...</span></span>
<span><span class="co">#&gt; ...                         ...                   ...           ...</span></span>
<span><span class="co">#&gt; 459                           1                     1              </span></span>
<span><span class="co">#&gt; 460                           1                     1              </span></span>
<span><span class="co">#&gt; 461                           1                     1              </span></span>
<span><span class="co">#&gt; 462               42;40;27;2...         12;10;0;1;... Zinc finge...</span></span>
<span><span class="co">#&gt; 463                           2                     1              </span></span>
<span><span class="co">#&gt;        Gene.names Fasta.headers Number.of.proteins  Peptides</span></span>
<span><span class="co">#&gt;       &lt;character&gt;   &lt;character&gt;          &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; 1      RBM8;Rbm8a &gt;tr|A0A023...                  3         1</span></span>
<span><span class="co">#&gt; 2            Dhx9 &gt;tr|A0A087...                  6        13</span></span>
<span><span class="co">#&gt; 3          Zmynd8 &gt;tr|Q3UH28...                  7         3</span></span>
<span><span class="co">#&gt; 4           Krt10 &gt;tr|A2A513...                  5        14</span></span>
<span><span class="co">#&gt; 5           Zmym4 &gt;sp|A2A791...                  3        56</span></span>
<span><span class="co">#&gt; ...           ...           ...                ...       ...</span></span>
<span><span class="co">#&gt; 459               &gt;sp|P28738...                  1         1</span></span>
<span><span class="co">#&gt; 460               &gt;sp|P58710...                  1         1</span></span>
<span><span class="co">#&gt; 461               &gt;sp|Q8BHW5...                  1         1</span></span>
<span><span class="co">#&gt; 462 Zfp280d;Zn... &gt;tr|V9GWW1...                 10        42</span></span>
<span><span class="co">#&gt; 463       Zfp280d &gt;tr|V9GXP2...                  1        29</span></span>
<span><span class="co">#&gt;     Razor.unique.peptides Unique.peptides Sequence.coverage</span></span>
<span><span class="co">#&gt;                 &lt;integer&gt;       &lt;integer&gt;         &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1                       1               1              10.9</span></span>
<span><span class="co">#&gt; 2                      13              13              13.1</span></span>
<span><span class="co">#&gt; 3                       3               3               3.6</span></span>
<span><span class="co">#&gt; 4                       1               1              18.7</span></span>
<span><span class="co">#&gt; 5                      56              56              48.7</span></span>
<span><span class="co">#&gt; ...                   ...             ...               ...</span></span>
<span><span class="co">#&gt; 459                     1               1               0.0</span></span>
<span><span class="co">#&gt; 460                     1               1               0.0</span></span>
<span><span class="co">#&gt; 461                     1               1               0.0</span></span>
<span><span class="co">#&gt; 462                    42              12              62.3</span></span>
<span><span class="co">#&gt; 463                     2               1              52.6</span></span>
<span><span class="co">#&gt;     Unique.razor.sequence.coverage Unique.sequence.coverage Mol.weight.kDa</span></span>
<span><span class="co">#&gt;                          &lt;numeric&gt;                &lt;numeric&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1                             10.9                     10.9         19.889</span></span>
<span><span class="co">#&gt; 2                             13.1                     13.1        149.620</span></span>
<span><span class="co">#&gt; 3                              3.6                      3.6        127.290</span></span>
<span><span class="co">#&gt; 4                              3.4                      3.4         57.040</span></span>
<span><span class="co">#&gt; 5                             48.7                     48.7        172.440</span></span>
<span><span class="co">#&gt; ...                            ...                      ...            ...</span></span>
<span><span class="co">#&gt; 459                            0.0                      0.0        109.270</span></span>
<span><span class="co">#&gt; 460                            0.0                      0.0         50.478</span></span>
<span><span class="co">#&gt; 461                            0.0                      0.0         36.570</span></span>
<span><span class="co">#&gt; 462                           62.3                     20.5        104.870</span></span>
<span><span class="co">#&gt; 463                            3.0                      1.5         90.174</span></span>
<span><span class="co">#&gt;     Sequence.length Sequence.lengths   Q.value     Score  Intensity      iBAQ</span></span>
<span><span class="co">#&gt;           &lt;integer&gt;      &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1               174    174;173;17... 0.0000000   15.2130   34384000   4298000</span></span>
<span><span class="co">#&gt; 2              1383    1383;1384;... 0.0000000   35.8400  156100000   2477800</span></span>
<span><span class="co">#&gt; 3              1154    1154;1174;... 0.0000000   44.9200   28924000    490240</span></span>
<span><span class="co">#&gt; 4               561    561;570;46... 0.0027174    4.5113   29053000   1117400</span></span>
<span><span class="co">#&gt; 5              1549    1549;1460;... 0.0000000  323.3100 2787500000  39822000</span></span>
<span><span class="co">#&gt; ...             ...              ...       ...       ...        ...       ...</span></span>
<span><span class="co">#&gt; 459             956              956 1.0000000   -2.0000   97210000       NaN</span></span>
<span><span class="co">#&gt; 460             440              440 0.0100760    3.6649 4734200000       NaN</span></span>
<span><span class="co">#&gt; 461             331              331 1.0000000   -2.0000  349630000       NaN</span></span>
<span><span class="co">#&gt; 462             949    949;974;78... 0.0000000  323.3100 3057000000  66457000</span></span>
<span><span class="co">#&gt; 463             799              799 0.0027322    4.6147   46397000   1104700</span></span>
<span><span class="co">#&gt;     MS.MS.Count Only.identified.by.site     Reverse Potential.contaminant</span></span>
<span><span class="co">#&gt;       &lt;integer&gt;             &lt;character&gt; &lt;character&gt;           &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1             7                                                          </span></span>
<span><span class="co">#&gt; 2            25                                                          </span></span>
<span><span class="co">#&gt; 3            10                                                          </span></span>
<span><span class="co">#&gt; 4             5                                                         +</span></span>
<span><span class="co">#&gt; 5           195                                                          </span></span>
<span><span class="co">#&gt; ...         ...                     ...         ...                   ...</span></span>
<span><span class="co">#&gt; 459           1                       +           +                      </span></span>
<span><span class="co">#&gt; 460          19                                   +                      </span></span>
<span><span class="co">#&gt; 461           5                       +           +                      </span></span>
<span><span class="co">#&gt; 462         148                                                          </span></span>
<span><span class="co">#&gt; 463           2                                                          </span></span>
<span><span class="co">#&gt;            id   Peptide.IDs Peptide.is.razor Mod.peptide.IDs  Evidence.IDs</span></span>
<span><span class="co">#&gt;     &lt;integer&gt;   &lt;character&gt;      &lt;character&gt;     &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1           0          2965             True            3291 21752;2175...</span></span>
<span><span class="co">#&gt; 2           1 13;191;192...    True;True;...   14;213;214... 33;34;35;3...</span></span>
<span><span class="co">#&gt; 3           2 1020;4113;...    True;True;...   1134;4574;... 6794;6795;...</span></span>
<span><span class="co">#&gt; 4           3 242;243;24...    False;Fals...   272;273;27... 1567;1568;...</span></span>
<span><span class="co">#&gt; 5           4 25;39;117;...    True;True;...   27;42;134;... 115;186;18...</span></span>
<span><span class="co">#&gt; ...       ...           ...              ...             ...           ...</span></span>
<span><span class="co">#&gt; 459       458          3542             True            3951 26674;2667...</span></span>
<span><span class="co">#&gt; 460       459           793             True             883 5390;5391;...</span></span>
<span><span class="co">#&gt; 461       460           171             True             191 1160;1161;...</span></span>
<span><span class="co">#&gt; 462       461 84;112;113...    True;True;...   98;128;129... 474;697;69...</span></span>
<span><span class="co">#&gt; 463       462 84;406;484...    False;Fals...   98;454;540... 474;2653;3...</span></span>
<span><span class="co">#&gt;         MS.MS.IDs    Best.MS.MS Oxidation.M.site.IDs Oxidation.M.site.positions</span></span>
<span><span class="co">#&gt;       &lt;character&gt;   &lt;character&gt;          &lt;character&gt;                &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1   37130;3713...         37131                                                </span></span>
<span><span class="co">#&gt; 2   68;69;70;2... 68;2489;24...                                                </span></span>
<span><span class="co">#&gt; 3   11288;1128... 11289;5402...                                                </span></span>
<span><span class="co">#&gt; 4   2851;2852;... 2879;2931;...                    0                        139</span></span>
<span><span class="co">#&gt; 5   158;244;24... 158;249;12...                                                </span></span>
<span><span class="co">#&gt; ...           ...           ...                  ...                        ...</span></span>
<span><span class="co">#&gt; 459         45839         45839                  509                        509</span></span>
<span><span class="co">#&gt; 460 9153;9154;...          9154                                                </span></span>
<span><span class="co">#&gt; 461 2371;2372;...          2375                  510                          4</span></span>
<span><span class="co">#&gt; 462 751;752;75... 751;1254;1...              511;512                      6;450</span></span>
<span><span class="co">#&gt; 463 751;752;75... 751;4530;5...                  512                        475</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="incorporating-information-about-the-experiment">Incorporating information about the experiment<a class="anchor" aria-label="anchor" href="#incorporating-information-about-the-experiment"></a>
</h2>
<p>In many cases it can be helpful to include additional metadata (that
may not be stored in the output files from the quantification tools)
about the experiment in the analysis report. This can be accommodated by
specifying the <code>experimentInfo</code> argument to the workflow
functions. This argument should be a named list, and all entries should
be scalar values. In the report, the list will be converted to a table
and displayed.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">experimentInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    `Experiment ID` <span class="op">=</span> <span class="st">"exp1"</span>,</span>
<span>    Description <span class="op">=</span> <span class="st">"A description of the experiment"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="defining-custom-ids-and-labels">Defining custom IDs and labels<a class="anchor" aria-label="anchor" href="#defining-custom-ids-and-labels"></a>
</h2>
<p><code>einprot</code> supports the definition of customized feature
IDs and labels, to accommodate different run modes of the quantification
tools and user perspectives. The identifiers are defined by the
following five arguments to the workflow functions:</p>
<ul>
<li>
<code>idCol</code> - defines the feature identifiers. These must be
unique.</li>
<li>
<code>labelCol</code> - defines the labels that will be used in most
plots (except where values are grouped by feature identifier, in which
case the <code>idCol</code> will be used instead to guarantee that
different features are not mixed).</li>
<li>
<code>geneidCol</code> - defines the gene symbol. This identifier
will be used to match features against pre-specified GO terms and
complexes, if requested. If no gene IDs are available, it can be set to
<code>NULL</code>.</li>
<li>
<code>proteinIdCol</code> - defines the protein identifiers. This
will be used to auto-generate links to UniProt and AlphaFold pages. It
should correspond to one or more UniProt identifiers, separated by
semi-colons.</li>
<li>
<code>stringIdCol</code> - defines the identifier that will be
matched against STRING. Any identifiers supported by STRING are allowed.
It can also be set to <code>NULL</code> if no STRING matching is
required.</li>
</ul>
<p>Each of these arguments can be either:</p>
<ul>
<li>a vector of annotation column names from the input text file (after
application of <code>make.names</code>). In this case, the values in the
provided columns will be concatenated to create the final values for the
corresponding feature identifiers.</li>
<li>a function with a single argument (a <code>data.frame</code>,
corresponding to the annotation columns of the input text file),
returning a character vector of identifiers.</li>
</ul>
<p>We illustrate how this works with a few examples. First, we read the
MaxQuant example data into a SingleCellExperiment object, as this
corresponds to how the data will be represented inside the
<code>einprot</code> workflows:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExperiment.html">importExperiment</a></span><span class="op">(</span><span class="va">mqFile</span>, iColPattern <span class="op">=</span> <span class="st">"^LFQ\\.intensity\\."</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">imp</span><span class="op">$</span><span class="va">sce</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SingleCellExperiment"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "SingleCellExperiment"</span></span></code></pre></div>
<p>The available annotation columns are stored in the
<code>rowData</code> of the SingleCellExperiment object:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Protein.IDs"                    "Majority.protein.IDs"          </span></span>
<span><span class="co">#&gt;  [3] "Peptide.counts.all"             "Peptide.counts.razor.unique"   </span></span>
<span><span class="co">#&gt;  [5] "Peptide.counts.unique"          "Protein.names"                 </span></span>
<span><span class="co">#&gt;  [7] "Gene.names"                     "Fasta.headers"                 </span></span>
<span><span class="co">#&gt;  [9] "Number.of.proteins"             "Peptides"                      </span></span>
<span><span class="co">#&gt; [11] "Razor.unique.peptides"          "Unique.peptides"               </span></span>
<span><span class="co">#&gt; [13] "Sequence.coverage"              "Unique.razor.sequence.coverage"</span></span>
<span><span class="co">#&gt; [15] "Unique.sequence.coverage"       "Mol.weight.kDa"                </span></span>
<span><span class="co">#&gt; [17] "Sequence.length"                "Sequence.lengths"              </span></span>
<span><span class="co">#&gt; [19] "Q.value"                        "Score"                         </span></span>
<span><span class="co">#&gt; [21] "Intensity"                      "iBAQ"                          </span></span>
<span><span class="co">#&gt; [23] "MS.MS.Count"                    "Only.identified.by.site"       </span></span>
<span><span class="co">#&gt; [25] "Reverse"                        "Potential.contaminant"         </span></span>
<span><span class="co">#&gt; [27] "id"                             "Peptide.IDs"                   </span></span>
<span><span class="co">#&gt; [29] "Peptide.is.razor"               "Mod.peptide.IDs"               </span></span>
<span><span class="co">#&gt; [31] "Evidence.IDs"                   "MS.MS.IDs"                     </span></span>
<span><span class="co">#&gt; [33] "Best.MS.MS"                     "Oxidation.M.site.IDs"          </span></span>
<span><span class="co">#&gt; [35] "Oxidation.M.site.positions"</span></span></code></pre></div>
<p>Let’s first define the <code>labelCol</code> (just as an example, the
procedure is identical for the other four arguments) as a vector of
column names:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">labelCol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene.names"</span>, <span class="st">"Majority.protein.IDs"</span><span class="op">)</span></span></code></pre></div>
<p>To see the effect this will have inside the <code>einprot</code>
workflows, we call the <code><a href="../reference/fixFeatureIds.html">fixFeatureIds()</a></code> function and ask it
to generate a new column in <code>rowData(sce)</code>, called
<code>einprotLabel</code> and defined by the <code>labelCol</code>
above. We see that the new column contains the values of the
<code>Gene.names</code> and <code>Majority.protein.IDs</code> columns,
concatenated.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Call the fixFeatureIds function to generate the new column</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixFeatureIds.html">fixFeatureIds</a></span><span class="op">(</span><span class="va">sce</span>, colDefs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>einprotLabel <span class="op">=</span> <span class="va">labelCol</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Compare the input columns (Gene.names and Majority.protein.IDs) to the </span></span>
<span><span class="co">## newly generate einprotLabel column, obtained by concatenating the two </span></span>
<span><span class="co">## input columns.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Gene.names</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RBM8;Rbm8a" "Dhx9"       "Zmynd8"     "Krt10"      "Zmym4"     </span></span>
<span><span class="co">#&gt; [6] "Rlf"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Majority.protein.IDs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A0A023T672;Q9CWZ3-2;Q9CWZ3"                       </span></span>
<span><span class="co">#&gt; [2] "A0A087WPL5;E9QNN1;O70133;O70133-2;O70133-3;Q3UR42"</span></span>
<span><span class="co">#&gt; [3] "Q3UH28;Q3U1M7;A2A483;E9Q8D1;A2A482;A2A484;A2A485" </span></span>
<span><span class="co">#&gt; [4] "A2A513;CON__P02535-1;P02535-3;P02535-2;P02535"    </span></span>
<span><span class="co">#&gt; [5] "A2A791;A2A791-2;F6VYE2"                           </span></span>
<span><span class="co">#&gt; [6] "A2A7F4;E9Q532"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">einprotLabel</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RBM8;Rbm8a.A0A023T672;Q9CWZ3-2;Q9CWZ3"                  </span></span>
<span><span class="co">#&gt; [2] "Dhx9.A0A087WPL5;E9QNN1;O70133;O70133-2;O70133-3;Q3UR42" </span></span>
<span><span class="co">#&gt; [3] "Zmynd8.Q3UH28;Q3U1M7;A2A483;E9Q8D1;A2A482;A2A484;A2A485"</span></span>
<span><span class="co">#&gt; [4] "Krt10.A2A513;CON__P02535-1;P02535-3;P02535-2;P02535"    </span></span>
<span><span class="co">#&gt; [5] "Zmym4.A2A791;A2A791-2;F6VYE2"                           </span></span>
<span><span class="co">#&gt; [6] "Rlf.A2A7F4;E9Q532"</span></span></code></pre></div>
<p>Next, let’s instead define the label as the first gene mentioned (for
each row) in the <code>Gene.names</code> column (before the first
semicolon). <code>einprot</code> provides a helper function for
this:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">labelCol</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="fu"><a href="../reference/getNthId.html">getFirstId</a></span><span class="op">(</span><span class="va">df</span>, colName <span class="op">=</span> <span class="st">"Gene.names"</span>, separator <span class="op">=</span> <span class="st">";"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixFeatureIds.html">fixFeatureIds</a></span><span class="op">(</span><span class="va">sce</span>, colDefs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>einprotLabel <span class="op">=</span> <span class="va">labelCol</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">einprotLabel</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RBM8"   "Dhx9"   "Zmynd8" "Krt10"  "Zmym4"  "Rlf"</span></span></code></pre></div>
<p><code>einprot</code> also provides a helper function for combining
multiple columns, possibly after splitting each of them by a given
separator and extracting the first value in each. We illustrate this by
defining the label as a combination of the gene name and the majority
protein ID, but only considering the first ID of each type:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">labelCol</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="fu"><a href="../reference/combineIds.html">combineIds</a></span><span class="op">(</span></span>
<span>    <span class="va">df</span>, combineCols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene.names"</span>, <span class="st">"Majority.protein.IDs"</span><span class="op">)</span>, </span>
<span>    combineWhen <span class="op">=</span> <span class="st">"always"</span>, splitSeparator <span class="op">=</span> <span class="st">";"</span>, </span>
<span>    joinSeparator <span class="op">=</span> <span class="st">"."</span>, makeUnique <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixFeatureIds.html">fixFeatureIds</a></span><span class="op">(</span><span class="va">sce</span>, colDefs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>einprotLabel <span class="op">=</span> <span class="va">labelCol</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">einprotLabel</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RBM8.A0A023T672" "Dhx9.A0A087WPL5" "Zmynd8.Q3UH28"   "Krt10.A2A513"   </span></span>
<span><span class="co">#&gt; [5] "Zmym4.A2A791"    "Rlf.A2A7F4"</span></span></code></pre></div>
<p>In the example above, the two columns were always combined. We can
also choose to only concatenate them when the first entry is not unique,
or when the first entry is missing (see the <code>combineWhen</code>
argument). We can also decide to require the returned values to be
unique (by setting <code>makeUnique</code> to <code>TRUE</code>).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">labelCol</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="fu"><a href="../reference/combineIds.html">combineIds</a></span><span class="op">(</span></span>
<span>    <span class="va">df</span>, combineCols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene.names"</span>, <span class="st">"Majority.protein.IDs"</span><span class="op">)</span>, </span>
<span>    combineWhen <span class="op">=</span> <span class="st">"nonunique"</span>, splitSeparator <span class="op">=</span> <span class="st">";"</span>, </span>
<span>    joinSeparator <span class="op">=</span> <span class="st">"."</span>, makeUnique <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixFeatureIds.html">fixFeatureIds</a></span><span class="op">(</span><span class="va">sce</span>, colDefs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>einprotLabel <span class="op">=</span> <span class="va">labelCol</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For most features, the gene name is unique</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">einprotLabel</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RBM8"   "Dhx9"   "Zmynd8" "Krt10"  "Zmym4"  "Rlf"</span></span>
<span></span>
<span><span class="co">## For some features, the protein IDs were added to the gene names to create </span></span>
<span><span class="co">## unique feature IDs</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">einprotLabel</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"\\."</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">einprotLabel</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "Ehmt1.A2AIS5"  "Ehmt2.A2CG76"  "Atrx.A6PWK7"   "Zfp462.B1AWL2"</span></span>
<span><span class="co">#&gt; [5] "Zfp462.B1AWL4" "Zfp462.B1AWL5"</span></span></code></pre></div>
<p>Finally, we define a custom function that splits the
<code>Fasta.headers</code> column and extracts the description of the
protein as the part between the first space and the first appearance of
the <code>OS=</code> pattern.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The column with fasta headers</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Fasta.headers</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "&gt;tr|A0A023T672|A0A023T672_MOUSE RNA-binding protein 8A OS=Mus musculus GN=RBM8 PE=1 SV=1;&gt;sp|Q9CWZ3-2|RBM8A_MOUSE Isoform 2 of RNA-binding protein 8A OS=Mus musculus GN=Rbm8a;&gt;sp|Q9CWZ3|RBM8A_MOUSE RNA-binding protein 8A OS=Mus musculus GN=Rbm8a PE=1 SV=3" </span></span>
<span><span class="co">#&gt; [2] "&gt;tr|A0A087WPL5|A0A087WPL5_MOUSE ATP-dependent RNA helicase A OS=Mus musculus GN=Dhx9 PE=1 SV=1;&gt;tr|E9QNN1|E9QNN1_MOUSE ATP-dependent RNA helicase A OS=Mus musculus GN=Dhx9 PE=1 SV=1;&gt;sp|O70133|DHX9_MOUSE ATP-dependent RNA helicase A OS=Mus musculus GN=Dhx9"</span></span>
<span><span class="co">#&gt; [3] "&gt;tr|Q3UH28|Q3UH28_MOUSE Protein Zmynd8 OS=Mus musculus GN=Zmynd8 PE=1 SV=1;&gt;tr|Q3U1M7|Q3U1M7_MOUSE Protein Zmynd8 OS=Mus musculus GN=Zmynd8 PE=1 SV=1;&gt;tr|A2A483|A2A483_MOUSE Protein Zmynd8 OS=Mus musculus GN=Zmynd8 PE=1 SV=1;&gt;tr|E9Q8D1|E9Q8D1_MOUSE Protein"</span></span>
<span><span class="co">#&gt; [4] "&gt;tr|A2A513|A2A513_MOUSE Keratin, type I cytoskeletal 10 OS=Mus musculus GN=Krt10 PE=3 SV=1;&gt;P02535-1 SWISS-PROT:P02535-1 Tax_Id=10090 Gene_Symbol=Krt10 Isoform 1 of Keratin, type I cytoskeletal 10;&gt;sp|P02535-3|K1C10_MOUSE Isoform 3 of Keratin, type I cytos"</span></span>
<span><span class="co">#&gt; [5] "&gt;sp|A2A791|ZMYM4_MOUSE Zinc finger MYM-type protein 4 OS=Mus musculus GN=Zmym4 PE=2 SV=1;&gt;sp|A2A791-2|ZMYM4_MOUSE Isoform 2 of Zinc finger MYM-type protein 4 OS=Mus musculus GN=Zmym4;&gt;tr|F6VYE2|F6VYE2_MOUSE Zinc finger MYM-type protein 4 (Fragment) OS=Mus "</span></span>
<span><span class="co">#&gt; [6] "&gt;tr|A2A7F4|A2A7F4_MOUSE Insulin-like 3 OS=Mus musculus GN=Rlf PE=4 SV=1;&gt;tr|E9Q532|E9Q532_MOUSE Insulin-like 3 OS=Mus musculus GN=Rlf PE=4 SV=1"</span></span>
<span></span>
<span><span class="co">## Define the custom function </span></span>
<span><span class="va">labelCol</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"[^ ]* (.*?) OS=.*"</span>, <span class="st">"\\1"</span>, <span class="va">df</span><span class="op">$</span><span class="va">Fasta.headers</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Generate the new column</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fixFeatureIds.html">fixFeatureIds</a></span><span class="op">(</span><span class="va">sce</span>, colDefs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>einprotLabel <span class="op">=</span> <span class="va">labelCol</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">einprotLabel</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RNA-binding protein 8A"          "ATP-dependent RNA helicase A"   </span></span>
<span><span class="co">#&gt; [3] "Protein Zmynd8"                  "Keratin, type I cytoskeletal 10"</span></span>
<span><span class="co">#&gt; [5] "Zinc finger MYM-type protein 4"  "Insulin-like 3"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="supported-column-patterns">Supported column patterns<a class="anchor" aria-label="anchor" href="#supported-column-patterns"></a>
</h2>
<p>The column pattern (provided to the <code>einprot</code> workflows
via the <code>iColPattern</code> argument) defines which columns of the
quantification file that will be used to generate the main assay in the
returned <code>SingleCellExperiment</code> object. Typically, for
<code>MaxQuant</code> output this would be either
<code>"^Intensity\\."</code>, <code>"^LFQ\\.intensity\\."</code> or
<code>"^iBAQ\\."</code>. It is also accepted to not escape the periods.
For <code>ProteomeDiscoverer</code>, the <code>iColPattern</code> is
typically either <code>"^Abundance\\.F.+\\.Sample\\."</code>,
<code>"^Abundances\\.Grouped\\."</code> or
<code>"^Abundance\\.F[0-9]+\\."</code>. For <code>FragPipe</code>, it is
usually <code>"\\.MaxLFQ\\.Intensity$"</code>. As mentioned above, it is
often helpful to first read the raw file into a
<code>SingleCellExperiment</code> object before launching the entire
workflow, to make sure that the column pattern is correctly
interpreted.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExperiment.html">importExperiment</a></span><span class="op">(</span><span class="va">mqFile</span>, iColPattern <span class="op">=</span> <span class="st">"^LFQ.intensity."</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExperiment.html">importExperiment</a></span><span class="op">(</span><span class="va">mqFile</span>, iColPattern <span class="op">=</span> <span class="st">"unknown_pattern"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in importExperiment(mqFile, iColPattern = "unknown_pattern"): Invalid iColPattern. Valid values: ^MS\.MS\.Count\., ^LFQ\.intensity\., ^Intensity\., ^Sequence\.coverage\., ^Unique\.peptides\., ^Razor\.+unique\.peptides\., ^Peptides\., ^iBAQ\., ^Top3\., ^Identification\.type\., ^Abundance\., ^Abundance\.F[0-9]+\., ^Abundance\.F.+\.Sample\., ^Abundances\.Count\., ^Abundances\.Count\.F[0-9]+\., ^Abundances\.Count\.F.+\.Sample\., ^Abundances\.Normalized\., ^Abundances\.Normalized\.F[0-9]+\., ^Abundances\.Normalized\.F.+\.Sample\., ^Abundances\.Grouped\.Count\., ^Abundances\.Grouped\.CV\.in\.Percent\., ^Abundances\.Grouped\., \.Unique\.Spectral\.Count$, \.Total\.Spectral\.Count$, \.Spectral\.Count$, \.Unique\.Intensity$, \.Total\.Intensity$, \.MaxLFQ\.Unique\.Intensity$, \.MaxLFQ\.Total\.Intensity$, \.MaxLFQ\.Intensity$, \.Intensity$, \.PG\.NrOfStrippedSequencesMeasured$, \.PG\.NrOfModifiedSequencesMeasured$, \.PG\.NrOfPrecursorsMeasured$, \.PG\.NrOfStrippedSequencesIdentified$, \.PG\.NrOfModifiedSequencesIdentified$, \.PG\.NrOfPrecursorsIdentified$, \.PG\.IsSingleHit$, \.PG\.NrOfStrippedSequencesUsedForQuantification$, \.PG\.NrOfModifiedSequencesUsedForQuantification$, \.PG\.NrOfPrecursorsUsedForQuantification$, \.PG\.Quantity$, \.PG\.Log2Quantity$, \.PG\.MS1Quantity$, \.PG\.MS2Quantity$, \.PG\.MS1ChannelQuantities$, \.PG\.MS2ChannelQuantities$, \.PG\.IBAQ$, \.PG\.ManuallyAccepted$, \.PG\.Cscore\.\.Run\.Wise\.$, \.PG\.QValue\.\.Run\.Wise\.$, \.PG\.PEP\.\.Run\.Wise\.$, \.PG\.PValue\.\.Run\.Wise\.$, \.PG\.IsIdentified$, \.PG\.RunEvidenceCount$, ^MS.MS.Count., ^LFQ.intensity., ^Intensity., ^Sequence.coverage., ^Unique.peptides., ^Razor.+unique.peptides., ^Peptides., ^iBAQ., ^Top3., ^Identification.type., ^Abundance., ^Abundance.F[0-9]+., ^Abundance.F.+.Sample., ^Abundances.Count., ^Abundances.Count.F[0-9]+., ^Abundances.Count.F.+.Sample., ^Abundances.Normalized., ^Abundances.Normalized.F[0-9]+., ^Abundances.Normalized.F.+.Sample., ^Abundances.Grouped.Count., ^Abundances.Grouped.CV.in.Percent., ^Abundances.Grouped., .Unique.Spectral.Count$, .Total.Spectral.Count$, .Spectral.Count$, .Unique.Intensity$, .Total.Intensity$, .MaxLFQ.Unique.Intensity$, .MaxLFQ.Total.Intensity$, .MaxLFQ.Intensity$, .Intensity$, .PG.NrOfStrippedSequencesMeasured$, .PG.NrOfModifiedSequencesMeasured$, .PG.NrOfPrecursorsMeasured$, .PG.NrOfStrippedSequencesIdentified$, .PG.NrOfModifiedSequencesIdentified$, .PG.NrOfPrecursorsIdentified$, .PG.IsSingleHit$, .PG.NrOfStrippedSequencesUsedForQuantification$, .PG.NrOfModifiedSequencesUsedForQuantification$, .PG.NrOfPrecursorsUsedForQuantification$, .PG.Quantity$, .PG.Log2Quantity$, .PG.MS1Quantity$, .PG.MS2Quantity$, .PG.MS1ChannelQuantities$, .PG.MS2ChannelQuantities$, .PG.IBAQ$, .PG.ManuallyAccepted$, .PG.Cscore..Run.Wise.$, .PG.QValue..Run.Wise.$, .PG.PEP..Run.Wise.$, .PG.PValue..Run.Wise.$, .PG.IsIdentified$, .PG.RunEvidenceCount$</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-sample-annotation-table">The sample annotation table<a class="anchor" aria-label="anchor" href="#the-sample-annotation-table"></a>
</h2>
<p>A sample annotation table must be provided when running each of the
<code>einprot</code> workflows. This <code>data.frame</code> must have
at least two columns, named <code>sample</code> and <code>group</code>,
but any additional columns are also supported and will be included in
the final <code>SingleCellExperiment</code> object. However, except for
the special case of a column named <code>batch</code> (see the section
on “Batch adjustment” below), additional columns will not be
automatically included when fitting the linear model.</p>
<p>The values in the <code>sample</code> column must correspond to the
column names of the imported <code>SingleCellExperiment</code> object,
which are generated by removing the specified column pattern from the
raw file column names. For example, in the data we imported in the
previous section, we have the following sample names:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">imp</span><span class="op">$</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Adnp_IP04"     "Adnp_IP05"     "Adnp_IP06"     "Chd4BF_IP07"  </span></span>
<span><span class="co">#&gt; [5] "Chd4BF_IP08"   "Chd4BF_IP09"   "RBC_ctrl_IP01" "RBC_ctrl_IP02"</span></span>
<span><span class="co">#&gt; [9] "RBC_ctrl_IP03"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="combining-groups-for-statistical-testing">Combining groups for statistical testing<a class="anchor" aria-label="anchor" href="#combining-groups-for-statistical-testing"></a>
</h2>
<p>By default, <code>einprot</code> will perform all pairwise
comparisons between the groups in the <code>group</code> column in the
sample annotation table (at least one of the groups in each comparison
must have at least two samples). In some cases, it is useful to specify
the comparisons explicitly. For example:</p>
<ul>
<li>We want to perform all pairwise comparisons, but we want to make
sure that a specific group is always interpreted as the ‘control group’
(the denominator in the comparison) whenever it is involved in a
comparison. This can be achieved by setting
<code>allPairwiseComparisons</code> to <code>TRUE</code>, and
<code>ctrlGroup</code> to the desired control group.</li>
<li>We only want to perform a subset of all possible pairwise
comparisons. In this case, we would directly specify
<code>comparisons</code> to be a list of character vector of length 2,
each defining a desired comparison (with the control level as the first
element). In the example data set above, we can define e.g. </li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparisons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RBC_ctrl"</span>, <span class="st">"Adnp"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RBC_ctrl"</span>, <span class="st">"Chd4BF"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>We want to combine multiple groups into one and consider all the
corresponding samples as replicates for some comparisons. This can be
achieved by defining the merged group in the <code>mergeGroup</code>
argument. The new group can then be used in the defined
comparisons:</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mergeGroup</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>newGroup <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Adnp"</span>, <span class="st">"Chd4BF"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">comparisons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RBC_ctrl"</span>, <span class="st">"newGroup"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>We want to compare a group against its complement (all samples not
in the group). This can be achieved by including ‘complement’ in the
definition of the comparison:</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparisons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"complement"</span>, <span class="st">"Adnp"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This will compare the <code>Adnp</code> group to all samples not in
the group.</p>
<p>It is also worth noting that if the <code>singleFit</code> argument
is set to <code>TRUE</code>, the interpretation of the comparisons
involving merged groups (and complements) changes somewhat. In this
case, instead of considering all samples in the merged group as
replicates, a model is first fit with all the original groups, and a
contrast is designed to compare the averages of the fitted values across
the merged groups in the comparison.</p>
</div>
<div class="section level2">
<h2 id="batch-adjustment-and-complex-designs">Batch adjustment and complex designs<a class="anchor" aria-label="anchor" href="#batch-adjustment-and-complex-designs"></a>
</h2>
<p><code>einprot</code> will adjust for batch effects by including an
extra term in the linear model (if the chosen statistical test is either
<code>limma</code> or <code>proDA</code>), if the sample annotation
table contains a column named <code>batch</code>. In this case, it will
also calculate a matrix of batch corrected abundance values, and use
this for visualization.</p>
<p>In the current version, all statistical tests in <code>einprot</code>
are based on pairwise comparisons of groups (possibly after including a
batch covariate in the model as described here). While this is likely to
cover a large fraction of practical use cases, there are setups that can
not be directly cast into this framework. A multi-factorial design
(e.g., samples from strain A and strain B, which are either treated or
untreated), can be accommodated by combining all predictors into a
single one (with values strainA-treated, strainA-untreated,
strainB-treated and strainB-untreated), and use this as the ‘group’
column, after which any pairwise comparisons between subgroups can be
performed. For more complex experimental designs, or to access the full
capabilities of statistical analysis packages like <code>limma</code>,
we recommend that users start with the exported
<code>SingleCellExperiment</code> object (<code>einprot</code> can be
run with <code>stattest</code> set to <code>"none"</code> to just
perform the processing, skipping the testing part), extract the
appropriate assay and the sample annotations (included in the
<code>colData</code> of the <code>SingleCellExperiment</code> object),
and set up the required test manually using a suitable package (such as
<code>limma</code>, <code>proDA</code>, or <code>prolfqua</code>).</p>
</div>
<div class="section level2">
<h2 id="valid-values-for-linktablecolumns-and-interactivedisplaycolumns">Valid values for <code>linkTableColumns</code> and
<code>interactiveDisplayColumns</code><a class="anchor" aria-label="anchor" href="#valid-values-for-linktablecolumns-and-interactivedisplaycolumns"></a>
</h2>
<p>The report generated by <code>einprot</code> contains several
interactive elements. If <code>addInteractiveVolcanos</code> is set to
<code>TRUE</code>, interactive volcano plots will be generated for each
comparison. In addition, a link table provides an overview of all the
features in the data set. The user can define additional columns to
include in this table, as well as columns to be displayed when hovering
over the points in the interactive volcano plots. Any column name (or
regular expression matching multiple column names) in the
<code>rowData</code> of the exported <code>SingleCellExperiment</code>
object can be displayed in the link table - to see what this entails for
a specific data set/quantification setup the easiest may be to run
through the workflow once, read the resulting
<code>SingleCellExperiment</code> and extract the column names of the
row data. For the interactive volcano plots, any column in the
statistical test result can be included, as well as any ‘annotation’
column in the row data of the <code>SingleCellExperiment</code> object
(i.e., not the columns corresponding to the test results that are copied
into the row data of the returned object). The separate test results are
included in the final <code>SingleCellExperiment</code> object and can
be accessed via</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testres</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">testres</span><span class="op">$</span><span class="va">tests</span></span></code></pre></div>
<p>(assuming that <code>sce</code> corresponds to the
<code>SingleCellExperiment</code> object returned by the
<code>einprot</code> workflows).</p>
</div>
<div class="section level2">
<h2 id="querying-the-metadata">Querying the metadata<a class="anchor" aria-label="anchor" href="#querying-the-metadata"></a>
</h2>
<p>In addition to reading the protein quantifications,
<code>einprot</code> also reads the metadata XML files generated by
<code>MaxQuant</code>, <code>Proteome Discoverer</code> and
<code>FragPipe</code>, and extracts information to display in the final
report. This information can also be converted into a nice-looking
table. The <code>Proteome Discoverer</code> summary function requires
the path to a folder that contains at least the
<code>_InputFiles.txt</code> and <code>_StudyInformation.txt</code>
output files from the <code>Proteome Discoverer</code> run, in addition
to the <code>.pdAnalysis</code> file with the run metadata (if one or
more of the files are missing, only the information from the available
file(s) is returned).</p>
<div class="section level3">
<h3 id="maxquant">MaxQuant<a class="anchor" aria-label="anchor" href="#maxquant"></a>
</h3>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## MaxQuant</span></span>
<span><span class="va">mqinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readMaxQuantXML.html">readMaxQuantXML</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mq_example"</span>, <span class="st">"1356_mqpar.xml"</span>, package <span class="op">=</span> <span class="st">"einprot"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/makeTableFromList.html">makeTableFromList</a></span><span class="op">(</span><span class="va">mqinfo</span><span class="op">)</span></span></code></pre></div>
<table class="table  lightable-paper lightable-striped" style='font-family: "Trebuchet MS", verdana, sans-serif; margin-left: auto; margin-right: auto;'><tbody>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
MaxQuant version
</td>
<td style="text-align:left;width: 100em; ">
1.5.3.8
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Parameter file
</td>
<td style="text-align:left;width: 100em; ">
/Users/runner/work/_temp/Library/einprot/extdata/mq_example/1356_mqpar.xml
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Search engine
</td>
<td style="text-align:left;width: 100em; ">
Andromeda
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Raw file location
</td>
<td style="text-align:left;width: 100em; ">
D:/Data/MaxQuant/
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Raw files
</td>
<td style="text-align:left;width: 100em; ">
F_160817_AdnpFB_IP04.raw, F_160817_AdnpFB_IP05.raw,
F_160817_AdnpFB_IP06.raw, F_160817_Chd4BF_IP07.raw,
F_160817_Chd4BF_IP08.raw, F_160817_Chd4BF_IP09.raw,
F_160817_RBC_ctrl_IP01.raw, F_160817_RBC_ctrl_IP02.raw,
F_160817_RBC_ctrl_IP03.raw
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Sample names
</td>
<td style="text-align:left;width: 100em; ">
Adnp_IP04, Adnp_IP05, Adnp_IP06, Chd4BF_IP07, Chd4BF_IP08, Chd4BF_IP09,
RBC_ctrl_IP01, RBC_ctrl_IP02, RBC_ctrl_IP03
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Databases
</td>
<td style="text-align:left;width: 100em; ">
D:/Databases/MOUSE__150129.fasta
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Contaminants
</td>
<td style="text-align:left;width: 100em; ">
*/MaxQuant_1.5.3.8/MaxQuant/bin/conf/contaminants.fasta
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Quantification settings (LFQ)
</td>
<td style="text-align:left;width: 100em; ">
LFQ min. ratio count: 1, fastLFQ: false, match-between runs (MBR): true,
Intensity based absolute quantification (iBAQ): true
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Min. razor peptides
</td>
<td style="text-align:left;width: 100em; ">
1
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Enzymes
</td>
<td style="text-align:left;width: 100em; ">
Trypsin/P
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Variable modifications
</td>
<td style="text-align:left;width: 100em; ">
Oxidation (M), Acetyl (Protein N-term)
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Fixed modifications
</td>
<td style="text-align:left;width: 100em; ">
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Max peptide mass
</td>
<td style="text-align:left;width: 100em; ">
8000
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Min peptide length
</td>
<td style="text-align:left;width: 100em; ">
7
</td>
</tr>
</tbody></table>
</div>
<div class="section level3">
<h3 id="proteome-discoverer">Proteome Discoverer<a class="anchor" aria-label="anchor" href="#proteome-discoverer"></a>
</h3>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Proteome Discoverer</span></span>
<span><span class="va">pdinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readProteomeDiscovererInfo.html">readProteomeDiscovererInfo</a></span><span class="op">(</span></span>
<span>    pdOutputFolder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"pdtmt_example"</span>, </span>
<span>                                 package <span class="op">=</span> <span class="st">"einprot"</span><span class="op">)</span>,</span>
<span>    pdResultName <span class="op">=</span> <span class="st">"Fig2_m23139_RTS_QC_varMods"</span>, </span>
<span>    pdAnalysisFile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"pdtmt_example"</span>, </span>
<span>                                 <span class="st">"Fig2_m23139_RTS_QC_varMods.pdAnalysis"</span>,</span>
<span>                                 package <span class="op">=</span> <span class="st">"einprot"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/makeTableFromList.html">makeTableFromList</a></span><span class="op">(</span><span class="va">pdinfo</span><span class="op">)</span></span></code></pre></div>
<table class="table  lightable-paper lightable-striped" style='font-family: "Trebuchet MS", verdana, sans-serif; margin-left: auto; margin-right: auto;'><tbody>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
PD version
</td>
<td style="text-align:left;width: 100em; ">
2.5.0.400
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
PD output folder
</td>
<td style="text-align:left;width: 100em; ">
/Users/runner/work/_temp/Library/einprot/extdata/pdtmt_example
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
PD result name
</td>
<td style="text-align:left;width: 100em; ">
Fig2_m23139_RTS_QC_varMods
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
PD analysis file
</td>
<td style="text-align:left;width: 100em; ">
/Users/runner/work/_temp/Library/einprot/extdata/pdtmt_example/Fig2_m23139_RTS_QC_varMods.pdAnalysis
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
PD Processing WF
</td>
<td style="text-align:left;width: 100em; ">
PWF_Tribrid_TMTpro_Quan_SPS_MS3_SequestHT_Percolator
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
PD Consensus WF
</td>
<td style="text-align:left;width: 100em; ">
CWF_Comprehensive_Enhanced Annotation_Reporter_Quan
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Search engine
</td>
<td style="text-align:left;width: 100em; ">
Sequest HT
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Instruments
</td>
<td style="text-align:left;width: 100em; ">
Orbitrap Fusion
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Raw file location
</td>
<td style="text-align:left;width: 100em; ">
D:/Data/PXD017803/
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Raw files
</td>
<td style="text-align:left;width: 100em; ">
Fig2_m23139_RTS.raw
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Sample names
</td>
<td style="text-align:left;width: 100em; ">
HIS4KO_S05, HIS4KO_S06, HIS4KO_S07, HIS4KO_S08, MET6KO_S01, MET6KO_S02,
MET6KO_S03, MET6KO_S04, URA2KO_S09, URA2KO_S10, URA2KO_S11, URA2KO_S12,
WT_S13, WT_S14, WT_S15, WT_S16
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Databases
</td>
<td style="text-align:left;width: 100em; ">
CON_iRT_contaminants_cRAPMaxQFMI_150507.fasta; YEAST__210503.fasta
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Contaminants
</td>
<td style="text-align:left;width: 100em; ">
CON_iRT_contaminants_cRAPMaxQFMI_150507.fasta
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Quantification settings (LFQ)
</td>
<td style="text-align:left;width: 100em; ">
Peptides used:Unique + Razor, quan. method: Reporter Ions Quantifier,
quan. MS order: MS3, abundance type: S/N, quan. correction: False, MS1
co-isolation threshold: 50, av. reporter SN threshold: 10, PSP mass
matches [%]: 65, norm. method: Total Peptide Amount, PD imputation: None
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Enzymes
</td>
<td style="text-align:left;width: 100em; ">
Trypsin (Full)
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Variable modifications
</td>
<td style="text-align:left;width: 100em; ">
Oxidation / +15.995 Da (M), Carbamidomethyl / +57.021 Da (C), TMTpro /
+304.207 Da (K, S, T), TMTpro / +304.207 Da (N-Terminus)
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Fixed modifications
</td>
<td style="text-align:left;width: 100em; ">
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Validation method
</td>
<td style="text-align:left;width: 100em; ">
PercolatorConfidenceAssignment
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Validation based on
</td>
<td style="text-align:left;width: 100em; ">
Target/Decoy, q-Value
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Confidence thresholds
</td>
<td style="text-align:left;width: 100em; ">
strict: 0.01, relaxed: 0.05
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Max missed cleavages
</td>
<td style="text-align:left;width: 100em; ">
2
</td>
</tr>
</tbody></table>
</div>
<div class="section level3">
<h3 id="fragpipe">FragPipe<a class="anchor" aria-label="anchor" href="#fragpipe"></a>
</h3>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## FragPipe</span></span>
<span><span class="va">fpinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFragPipeInfo.html">readFragPipeInfo</a></span><span class="op">(</span></span>
<span>    fragpipeDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"fp_example"</span>,</span>
<span>                              package <span class="op">=</span> <span class="st">"einprot"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/makeTableFromList.html">makeTableFromList</a></span><span class="op">(</span><span class="va">fpinfo</span><span class="op">)</span></span></code></pre></div>
<table class="table  lightable-paper lightable-striped" style='font-family: "Trebuchet MS", verdana, sans-serif; margin-left: auto; margin-right: auto;'><tbody>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
FragPipe version
</td>
<td style="text-align:left;width: 100em; ">
19.1
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
FragPipe parameter file
</td>
<td style="text-align:left;width: 100em; ">
/Users/runner/work/_temp/Library/einprot/extdata/fp_example/fragpipe.workflow
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
FragPipe log file
</td>
<td style="text-align:left;width: 100em; ">
/Users/runner/work/_temp/Library/einprot/extdata/fp_example/log_2023-04-12_20-12-46.txt
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Search engine
</td>
<td style="text-align:left;width: 100em; ">
MSFragger-3.7
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Raw file location
</td>
<td style="text-align:left;width: 100em; ">
D:/Data/FUSION
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Raw files
</td>
<td style="text-align:left;width: 100em; ">
F_160817_AdnpFB_IP06.raw, F_160817_AdnpFB_IP05.raw,
F_160817_RBC_ctrl_IP02.raw, F_160817_AdnpFB_IP04.raw,
F_160817_RBC_ctrl_IP01.raw, F_160817_Chd4BF_IP09.raw,
F_160817_Chd4BF_IP07.raw, F_160817_Chd4BF_IP08.raw,
F_160817_RBC_ctrl_IP03.raw
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Sample names
</td>
<td style="text-align:left;width: 100em; ">
Adnp_IP04, Adnp_IP05, Adnp_IP06, Chd4BF_IP07, Chd4BF_IP08, Chd4BF_IP09,
RBC_ctrl_IP01, RBC_ctrl_IP02, RBC_ctrl_IP03
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Databases
</td>
<td style="text-align:left;width: 100em; ">
D/://Data//FASTA//2023-04-12-decoys-contam_MOUSE__190410.fasta.fas
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Peptides (ranges)
</td>
<td style="text-align:left;width: 100em; ">
length: 7-50 AA; mass: 500-5000 Da
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Mass error tolerances
</td>
<td style="text-align:left;width: 100em; ">
precursor:-20-20 [ppm]; fragment:0.7 [Da] (after optimization:200 PPM)
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Quantification settings (LFQ)
</td>
<td style="text-align:left;width: 100em; ">
IonQuant: TRUE, Calculate MaxLFQ intensity: TRUE, Normalization: TRUE,
match-between runs (MBR): FALSE, min. ions: 2, Top N ions: 3
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Enzymes
</td>
<td style="text-align:left;width: 100em; ">
stricttrypsin[KR, C-terminal, 2 missed cleavages]
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Variable modifications
</td>
<td style="text-align:left;width: 100em; ">
M(15.9949), N-term(42.0106)
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Fixed modifications
</td>
<td style="text-align:left;width: 100em; ">
C(57.0215)
</td>
</tr>
<tr>
<td style="text-align:left;width: 25em; font-weight: bold;border-right:1px solid;">
Database decoy tag
</td>
<td style="text-align:left;width: 100em; ">
rev_
</td>
</tr>
</tbody></table>
</div>
</div>
<div class="section level2">
<h2 id="frequently-asked-questions">Frequently asked questions<a class="anchor" aria-label="anchor" href="#frequently-asked-questions"></a>
</h2>
<div class="section level3">
<h3 id="what-is-a-singlecellexperiment-object-and-why-do-we-use-it-for-proteomics-data">What is a <code>SingleCellExperiment</code> object, and why do we
use it for proteomics data?<a class="anchor" aria-label="anchor" href="#what-is-a-singlecellexperiment-object-and-why-do-we-use-it-for-proteomics-data"></a>
</h3>
<p>The <a href="https://bioconductor.org/packages/SingleCellExperiment/" class="external-link">SingleCellExperiment</a>
class is a standard data container in the <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> project. It provides
an extension of the <a href="https://bioconductor.org/packages/SummarizedExperiment/" class="external-link">SummarizedExperiment</a>
container. Despite the name, it is equally suitable to use for storing
many different types of rectangular data sets, including those
considered in <code>einprot</code>. The advantage compared to the
<code>SummarizedExperiment</code> object, for our purposes, is that in
addition to abundance values and annotations for samples and features,
it can also hold low-dimensional representations, here obtained by PCA.
Furthermore, using a standard container makes it possible to directly
use the <code>einprot</code> output as input to functions from a variety
of Bioconductor packages.</p>
</div>
<div class="section level3">
<h3 id="which-statistical-test-frameworks-are-supported-by-einprot">Which statistical test frameworks are supported by
<code>einprot</code>?<a class="anchor" aria-label="anchor" href="#which-statistical-test-frameworks-are-supported-by-einprot"></a>
</h3>
<p><code>einprot</code> provides access to three different statistical
frameworks. The default is to use the <code>limma</code> package <span class="citation">(Ritchie et al. 2015)</span>, which has been shown to
perform well for analysis of proteomics data <span class="citation">(Peng et al. 2023)</span>. <code>limma</code> is a
general-purpose inference package that fits a linear model and performs
inference based on a moderated t-statistic. It can accommodate batch
effects as well as individual sample weights, if required. We also
provide the option of performing a t-test, which more closely mimics the
default setup of <code>Perseus</code> <span class="citation">(Tyanova et
al. 2016)</span>. Note that in this case, no batch effect adjustment can
be made, and some plots will not be generated. Finally, we provide the
option to use <code>proDA</code> <span class="citation">(Ahlmann-Eltze
and Anders 2020)</span>, which is a statistical analysis package
developed specifically for proteomics data. The main feature here is
that no imputation of missing values is required; this will instead be
accounted for internally by a probabilistic dropout model.</p>
<p>For <code>limma</code> and <code>proDA</code>, <code>einprot</code>
offers the possibility of either fitting a single model to all samples,
and extract comparisons of interest using linear contrasts, or to subset
the data to only the samples used for each comparison. The main
advantage of fitting a single model (<code>singleFit = TRUE</code>) is
that a larger number of samples are used to estimate parameters, which
usually give more precise estimates. For this reason, this is typically
the recommended approach for <code>limma</code> and other inference
pipelines. However, it also involves making assumptions about
similarities of variances between groups, and if there are large
differences, either fitting separate models or potentially using a
weighting approach may be more suitable. For more discussions about this
topic, see e.g. the following posts from the Bioconductor support forum:
<a href="https://support.bioconductor.org/p/60556/" class="external-link uri">https://support.bioconductor.org/p/60556/</a>, <a href="https://support.bioconductor.org/p/88032/" class="external-link uri">https://support.bioconductor.org/p/88032/</a>, <a href="https://support.bioconductor.org/p/61556/" class="external-link uri">https://support.bioconductor.org/p/61556/</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><summary><b> Click to expand </b>
</summary><pre><code><span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.7.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] einprot_0.9.7               SummarizedExperiment_1.36.0</span></span>
<span><span class="co">#&gt;  [3] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">#&gt;  [5] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">#&gt;  [7] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt;  [9] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] STRINGdb_2.18.0             fs_1.6.5                   </span></span>
<span><span class="co">#&gt;   [3] ProtGenerics_1.38.0         bitops_1.0-9               </span></span>
<span><span class="co">#&gt;   [5] DirichletMultinomial_1.48.0 TFBSTools_1.44.0           </span></span>
<span><span class="co">#&gt;   [7] httr_1.4.7                  ash_1.0-15                 </span></span>
<span><span class="co">#&gt;   [9] RColorBrewer_1.1-3          doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [11] tools_4.4.3                 R6_2.6.1                   </span></span>
<span><span class="co">#&gt;  [13] DT_0.33                     lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt;  [15] mgcv_1.9-1                  GetoptLong_1.0.5           </span></span>
<span><span class="co">#&gt;  [17] withr_3.0.2                 iSEEhex_1.8.0              </span></span>
<span><span class="co">#&gt;  [19] gridExtra_2.3               GGally_2.2.1               </span></span>
<span><span class="co">#&gt;  [21] ggalt_0.4.0                 cli_3.6.4                  </span></span>
<span><span class="co">#&gt;  [23] textshaping_1.0.0           shinyjs_2.1.0              </span></span>
<span><span class="co">#&gt;  [25] sandwich_3.1-1              sass_0.4.9                 </span></span>
<span><span class="co">#&gt;  [27] robustbase_0.99-4-1         mvtnorm_1.3-3              </span></span>
<span><span class="co">#&gt;  [29] readr_2.1.5                 genefilter_1.88.0          </span></span>
<span><span class="co">#&gt;  [31] pkgdown_2.1.1.9000          Rsamtools_2.22.0           </span></span>
<span><span class="co">#&gt;  [33] systemfonts_1.2.1           R.utils_2.13.0             </span></span>
<span><span class="co">#&gt;  [35] svglite_2.1.3               stringdist_0.9.15          </span></span>
<span><span class="co">#&gt;  [37] scater_1.34.1               rrcov_1.7-6                </span></span>
<span><span class="co">#&gt;  [39] plotrix_3.8-4               BSgenome_1.74.0            </span></span>
<span><span class="co">#&gt;  [41] maps_3.4.2.1                limma_3.62.2               </span></span>
<span><span class="co">#&gt;  [43] rstudioapi_0.17.1           impute_1.80.0              </span></span>
<span><span class="co">#&gt;  [45] RSQLite_2.3.9               BiocIO_1.16.0              </span></span>
<span><span class="co">#&gt;  [47] generics_0.1.3              shape_1.4.6.1              </span></span>
<span><span class="co">#&gt;  [49] gtools_3.9.5                dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [51] GO.db_3.20.0                Matrix_1.7-3               </span></span>
<span><span class="co">#&gt;  [53] ggbeeswarm_0.7.2            imputeLCMD_2.1             </span></span>
<span><span class="co">#&gt;  [55] abind_1.4-8                 R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt;  [57] lifecycle_1.0.4             yaml_2.3.10                </span></span>
<span><span class="co">#&gt;  [59] iSEEu_1.18.0                gplots_3.2.0               </span></span>
<span><span class="co">#&gt;  [61] SparseArray_1.6.2           grid_4.4.3                 </span></span>
<span><span class="co">#&gt;  [63] blob_1.2.4                  promises_1.3.2             </span></span>
<span><span class="co">#&gt;  [65] pwalign_1.2.0               crayon_1.5.3               </span></span>
<span><span class="co">#&gt;  [67] shinydashboard_0.7.2        miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt;  [69] lattice_0.22-6              msigdbr_10.0.0             </span></span>
<span><span class="co">#&gt;  [71] beachmat_2.22.0             ComplexUpset_1.3.3         </span></span>
<span><span class="co">#&gt;  [73] cowplot_1.1.3               annotate_1.84.0            </span></span>
<span><span class="co">#&gt;  [75] KEGGREST_1.46.0             pillar_1.10.1              </span></span>
<span><span class="co">#&gt;  [77] knitr_1.49                  ComplexHeatmap_2.22.0      </span></span>
<span><span class="co">#&gt;  [79] rjson_0.2.23                codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [81] glue_1.8.0                  ggiraph_0.8.12             </span></span>
<span><span class="co">#&gt;  [83] data.table_1.17.0           pcaMethods_1.98.0          </span></span>
<span><span class="co">#&gt;  [85] MultiAssayExperiment_1.32.0 vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [87] png_0.1-8                   poweRlaw_1.0.0             </span></span>
<span><span class="co">#&gt;  [89] gtable_0.3.6                gsubfn_0.7                 </span></span>
<span><span class="co">#&gt;  [91] cachem_1.1.0                xfun_0.51                  </span></span>
<span><span class="co">#&gt;  [93] S4Arrays_1.6.0              mime_0.12                  </span></span>
<span><span class="co">#&gt;  [95] pcaPP_2.0-5                 survival_3.8-3             </span></span>
<span><span class="co">#&gt;  [97] rrcovNA_0.5-2               SingleCellExperiment_1.28.1</span></span>
<span><span class="co">#&gt;  [99] iterators_1.0.14            gmm_1.8                    </span></span>
<span><span class="co">#&gt; [101] iSEE_2.18.0                 statmod_1.5.0              </span></span>
<span><span class="co">#&gt; [103] nlme_3.1-167                bit64_4.6.0-1              </span></span>
<span><span class="co">#&gt; [105] ExploreModelMatrix_1.18.0   bslib_0.9.0                </span></span>
<span><span class="co">#&gt; [107] tmvtnorm_1.6                irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt; [109] vipor_0.4.7                 KernSmooth_2.23-26         </span></span>
<span><span class="co">#&gt; [111] seqLogo_1.72.0              colorspace_2.1-1           </span></span>
<span><span class="co">#&gt; [113] DBI_1.2.3                   ade4_1.7-23                </span></span>
<span><span class="co">#&gt; [115] proDA_1.20.0                motifStack_1.50.0          </span></span>
<span><span class="co">#&gt; [117] tidyselect_1.2.1            curl_6.2.1                 </span></span>
<span><span class="co">#&gt; [119] bit_4.6.0                   compiler_4.4.3             </span></span>
<span><span class="co">#&gt; [121] extrafontdb_1.0             chron_2.3-62               </span></span>
<span><span class="co">#&gt; [123] BiocNeighbors_2.0.1         xml2_1.3.8                 </span></span>
<span><span class="co">#&gt; [125] plotly_4.10.4               desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [127] DelayedArray_0.32.0         rtracklayer_1.66.0         </span></span>
<span><span class="co">#&gt; [129] colourpicker_1.3.0          msigdbdf_24.1.0            </span></span>
<span><span class="co">#&gt; [131] scales_1.3.0                caTools_1.18.3             </span></span>
<span><span class="co">#&gt; [133] DEoptimR_1.1-3-1            proj4_1.0-14               </span></span>
<span><span class="co">#&gt; [135] hexbin_1.28.5               stringr_1.5.1              </span></span>
<span><span class="co">#&gt; [137] digest_0.6.37               rmarkdown_2.29             </span></span>
<span><span class="co">#&gt; [139] XVector_0.46.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [141] pkgconfig_2.0.3             extrafont_0.19             </span></span>
<span><span class="co">#&gt; [143] listviewer_4.0.0            fastmap_1.2.0              </span></span>
<span><span class="co">#&gt; [145] rlang_1.1.5                 GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt; [147] htmlwidgets_1.6.4           UCSC.utils_1.2.0           </span></span>
<span><span class="co">#&gt; [149] shiny_1.10.0                farver_2.1.2               </span></span>
<span><span class="co">#&gt; [151] jquerylib_0.1.4             zoo_1.8-13                 </span></span>
<span><span class="co">#&gt; [153] jsonlite_1.9.1              BiocParallel_1.40.0        </span></span>
<span><span class="co">#&gt; [155] mclust_6.1.1                R.oo_1.27.0                </span></span>
<span><span class="co">#&gt; [157] RCurl_1.98-1.16             BiocSingular_1.22.0        </span></span>
<span><span class="co">#&gt; [159] magrittr_2.0.3              scuttle_1.16.0             </span></span>
<span><span class="co">#&gt; [161] kableExtra_1.4.0            GenomeInfoDbData_1.2.13    </span></span>
<span><span class="co">#&gt; [163] patchwork_1.3.0             munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [165] Rcpp_1.0.14                 viridis_0.6.5              </span></span>
<span><span class="co">#&gt; [167] babelgene_22.9              proto_1.0.0                </span></span>
<span><span class="co">#&gt; [169] MsCoreUtils_1.18.0          sqldf_0.4-11               </span></span>
<span><span class="co">#&gt; [171] stringi_1.8.4               rintrojs_0.3.4             </span></span>
<span><span class="co">#&gt; [173] zlibbioc_1.52.0             MASS_7.3-65                </span></span>
<span><span class="co">#&gt; [175] plyr_1.8.9                  ggstats_0.9.0              </span></span>
<span><span class="co">#&gt; [177] parallel_4.4.3              ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt; [179] forcats_1.0.0               CNEr_1.42.0                </span></span>
<span><span class="co">#&gt; [181] Biostrings_2.74.1           splines_4.4.3              </span></span>
<span><span class="co">#&gt; [183] hash_2.2.6.3                hms_1.1.3                  </span></span>
<span><span class="co">#&gt; [185] circlize_0.4.16             igraph_2.1.4               </span></span>
<span><span class="co">#&gt; [187] uuid_1.2-1                  QFeatures_1.16.0           </span></span>
<span><span class="co">#&gt; [189] reshape2_1.4.4              ScaledMatrix_1.14.0        </span></span>
<span><span class="co">#&gt; [191] TFMPvalue_0.0.9             XML_3.99-0.18              </span></span>
<span><span class="co">#&gt; [193] evaluate_1.0.3              tzdb_0.4.0                 </span></span>
<span><span class="co">#&gt; [195] foreach_1.5.2               httpuv_1.6.15              </span></span>
<span><span class="co">#&gt; [197] Rttf2pt1_1.3.12             tidyr_1.3.1                </span></span>
<span><span class="co">#&gt; [199] purrr_1.0.4                 clue_0.3-66                </span></span>
<span><span class="co">#&gt; [201] norm_1.0-11.1               ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt; [203] rsvd_1.0.5                  xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [205] restfulr_0.0.15             AnnotationFilter_1.30.0    </span></span>
<span><span class="co">#&gt; [207] later_1.4.1                 viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [209] ragg_1.3.3                  tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [211] beeswarm_0.4.0              GenomicAlignments_1.42.0   </span></span>
<span><span class="co">#&gt; [213] memoise_2.0.1               AnnotationDbi_1.68.0       </span></span>
<span><span class="co">#&gt; [215] writexl_1.5.1               cluster_2.1.8.1            </span></span>
<span><span class="co">#&gt; [217] shinyWidgets_0.9.0          shinyAce_0.4.4</span></span></code></pre>
</details>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-AhlmannEltze2020proda" class="csl-entry">
Ahlmann-Eltze, Constantin, and Simon Anders. 2020. <span>“<span class="nocase">proDA</span>: Probabilistic Dropout Analysis for
Identifying Differentially Abundant Proteins in <span>Label-Free</span>
Mass Spectrometry.”</span> <em>bioRxiv
Doi:https://Doi.org/10.1101/661496</em>.
</div>
<div id="ref-Cox2008maxquant" class="csl-entry">
Cox, J, and M Mann. 2008. <span>“<span>MaxQuant</span> Enables High
Peptide Identification Rates, Individualized p.p.b.-Range Mass
Accuracies and Proteome-Wide Protein Quantification.”</span> <em>Nature
Biotechnology</em> 26 (12): 1367–72. <a href="https://www.nature.com/articles/nbt.1511" class="external-link">https://www.nature.com/articles/nbt.1511</a>.
</div>
<div id="ref-Orsburn2021pd" class="csl-entry">
Orsburn, BC. 2021. <span>“Proteome <span>Discoverer-A</span> Community
Enhanced Data Processing Suite for Protein Informatics.”</span>
<em>Proteomes</em> 9 (1). <a href="https://www.mdpi.com/2227-7382/9/1/15" class="external-link">https://www.mdpi.com/2227-7382/9/1/15</a>.
</div>
<div id="ref-Ostapcuk2018chd4" class="csl-entry">
Ostapcuk, Veronika, Fabio Mohn, Sarah H Carl, Anja Basters, Daniel Hess,
Vytautas Iesmantavicius, Lisa Lampersberger, et al. 2018.
<span>“Activity-Dependent Neuroprotective Protein Recruits
<span>HP1</span> and <span>CHD4</span> to Control Lineage-Specifying
Genes.”</span> <em>Nature</em> 557 (7707): 739–43.
</div>
<div id="ref-Peng2023optim" class="csl-entry">
Peng, Hui, He Wang, Weijia Kong, Jinyan Li, and Wilson Wen Bin Goh.
2023. <span>“Optimizing Proteomics Data Differential Expression Analysis
via <span>High-Performing</span> Rules and Ensemble Inference.”</span>
<em>bioRxiv Doi:10.1101/2023.06.26.546625</em>. <a href="https://doi.org/10.1101/2023.06.26.546625" class="external-link">https://doi.org/10.1101/2023.06.26.546625</a>.
</div>
<div id="ref-Ritchie2015limma" class="csl-entry">
Ritchie, ME, B Phipson, D Wu, Y Hu, CW Law, W Shi, and GK Smyth. 2015.
<span>“Limma Powers Differential Expression Analyses for <span class="nocase">RNA-sequencing</span> and Microarray Studies.”</span>
<em>Nucleic Acids Research</em> 43 (7): e47. <a href="https://academic.oup.com/nar/article/43/7/e47/2414268" class="external-link">https://academic.oup.com/nar/article/43/7/e47/2414268</a>.
</div>
<div id="ref-Rue-Albrecht2018isee" class="csl-entry">
Rue-Albrecht, Kevin, Federico Marini, Charlotte Soneson, and Aaron T L
Lun. 2018. <span>“<span class="nocase">iSEE</span>: Interactive
<span>SummarizedExperiment</span> Explorer.”</span> <em>F1000Res.</em>
7: 741.
</div>
<div id="ref-Tyanova2016perseus" class="csl-entry">
Tyanova, S, T Temu, P Sinitcyn, A Carlson, MY Hein, T Geiger, M Mann,
and J Cox. 2016. <span>“The Perseus Computational Platform for
Comprehensive Analysis of (Prote)omics Data.”</span> <em>Nature
Methods</em> 13 (9): 731–40. <a href="https://www.nature.com/articles/nmeth.3901" class="external-link">https://www.nature.com/articles/nmeth.3901</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Charlotte Soneson, Jan Seebacher, Michael Stadler.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
